---
title: "scRNAseq Captured Genes: Mouse Lung (with replicates)"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction

## Objective
In this Rmarkdown document, we are going to 

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(ggrepel)
library(UpSetR)
library(gt)
library(grid)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
sub_folder <- "SCGTEST_32"
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_r_objects <- here::here("02_QC/results/R_objects")

# Functions
source(here::here("bin/utils.R"))

fromList_togetdata <- function (input) {
  # Same as original fromList()...
  elements <- unique(unlist(input))
  data <- unlist(lapply(input, function(x) {
      x <- as.vector(match(elements, x))
      }))
  data[is.na(data)] <- as.integer(0)
  data[data != 0] <- as.integer(1)
  data <- data.frame(matrix(data, ncol = length(input), byrow = F))
  data <- data[which(rowSums(data) != 0), ]
  names(data) <- names(input)
  # ... Except now it conserves your original value names!
  row.names(data) <- elements
  return(data)
  }
```


## Load data
The data used in this Rmarkdown document comes from **SCGTEST_32** dataset, filtered matrices were processed with `cellranger v6.1.1`, and the doublet predictions were obtained using `scrublet`.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)
subproject_folder_n <- "MouseLung_fixed"

# Merging metadata tables
metadata <- metadata %>% 
  filter((subproject == "SCGTEST_32" | subproject == "SCGTEST_33")
         & tissue == "Lung"
         & ( sample_protocol == "Fresh" | sample_protocol == "Fix&Cut_24h"))

# Modify metadata
metadata$library_name[metadata$library_name == "Lung_fresh_0"] <- "Fresh_1"
metadata$library_name[metadata$library_name == "Lung_fresh"] <- "Fresh_2"
metadata$library_name[metadata$library_name == "Lung_fixed_24h_0"] <- "Fixed_1"
metadata$library_name[metadata$library_name == "Lung_fixed_24h"] <- "Fixed_2"

metadata$sample_protocol[metadata$library_name == "Fresh_1" | metadata$library_name == "Fresh_2"] <- "Fresh"
metadata$sample_protocol[metadata$library_name == "Fixed_1" | metadata$library_name == "Fixed_2"] <- "Fixed"

metadata$replicate[metadata$library_name == "Fresh_1" | metadata$library_name == "Fixed_1"] <- "Replicate 1"
metadata$replicate[metadata$library_name == "Fresh_2" | metadata$library_name == "Fixed_2"] <- "Replicate 2"


print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))

# Load object data -- from downsampled reads
seurat_obj <- readRDS(str_c(path_r_objects, "/FIXnCUT_MouseLung_fixed_merged.rds"))
seurat_obj <- Seurat::NormalizeData(seurat_obj,
                                    normalization.method = "LogNormalize",
                                    scale.factor = 10000,
                                    margin = 1
                                    )

# Removing all genes that have expression 0
n_cells <- Matrix::rowSums(seurat_obj[["RNA"]]@counts > 0)
keep_genes <- rownames(seurat_obj)[n_cells > 0]
seurat_obj <- subset(seurat_obj, features = keep_genes)
```


# Exclusive genes by library
```{r}
seurat_obj_list <- SplitObject(seurat_obj, split.by = "library_name")


seurat_obj_list <- purrr::map(seurat_obj_list, function(seurat) {
  # Number of cells expressing each gene, excluding 0 counts
  n_cells <- Matrix::rowSums(seurat[["RNA"]]@counts > 0)
  
  # Subset genes that are expressed in data
  keep_genes <- rownames(seurat)[n_cells > 0]

  # Filtering out genes expressed in few cells
  seurat_obj <- subset(seurat, features = keep_genes)
  seurat_obj  
})
seurat_obj_list
```

## Upset Plot
https://github.com/hms-dbmi/UpSetR

```{r fig.height=5, fig.width=12}
UpSetR::upset(
  UpSetR::fromList(list(
    "Fresh_1" = rownames(seurat_obj_list$Fresh_1),
    "Fresh_2" = rownames(seurat_obj_list$Fresh_2),
    "Fixed_1" = rownames(seurat_obj_list$Fixed_1),
    "Fixed_2" = rownames(seurat_obj_list$Fixed_2)
  )),
  order.by = "freq",
  #group.by = "sets",
  mainbar.y.label = "# Genes", sets.x.label = "# Genes / sample",
  point.size = 5, line.size = 1, text.scale = 2,
  empty.intersections = "on",
   queries = list(
     list(query = intersects, params = list("Fixed_1", "Fixed_2", "Fresh_1", "Fresh_2"), color = "blue", active = T),
     list(query = intersects, params = list("Fixed_1", "Fixed_2"), color = "orange", active = T),
     list(query = intersects, params = list("Fresh_1", "Fresh_2"), color = "green", active = T),
     list(query = intersects, params = list("Fresh_1"), color = "green", active = T),
     list(query = intersects, params = list("Fresh_2"), color = "green", active = T),
     list(query = intersects, params = list("Fixed_1"), color = "orange", active = T),
     list(query = intersects, params = list("Fixed_2"), color = "orange", active = T)
     )
  )
```

## Manual exploration
```{r}
genes_lib_df <- fromList_togetdata(list(
    "Fresh_1" = rownames(seurat_obj_list$Fresh_1),
    "Fresh_2" = rownames(seurat_obj_list$Fresh_2),
    "Fixed_1" = rownames(seurat_obj_list$Fixed_1),
    "Fixed_2" = rownames(seurat_obj_list$Fixed_2)
  ))
genes_lib_df <- genes_lib_df[order(row.names(genes_lib_df)), ]
  
head(genes_lib_df)
```

# Exclusive genes by protocol
```{r}
seurat_obj_list_prot <- SplitObject(seurat_obj, split.by = "sample_protocol")


seurat_obj_list_prot <- purrr::map(seurat_obj_list_prot, function(seurat) {
  # Number of cells expressing each gene, excluding 0 counts
  n_cells <- Matrix::rowSums(seurat[["RNA"]]@counts > 0)
  
  # Subset genes that are expressed in data
  keep_genes <- rownames(seurat)[n_cells > 0]

  # Filtering out genes expressed in few cells
  seurat_obj <- subset(seurat, features = keep_genes)
  seurat_obj  
})
seurat_obj_list_prot
```

## Upset Plot
https://github.com/hms-dbmi/UpSetR

```{r fig.height=5, fig.width=8}
UpSetR::upset(
  UpSetR::fromList(list(
    "Fresh" = rownames(seurat_obj_list_prot$Fresh),
    "Fixed" = rownames(seurat_obj_list_prot$Fixed)
  )),
  order.by = "freq",
  #group.by = "sets",
  mainbar.y.label = "# Genes", sets.x.label = "# Genes / condition",
  point.size = 4, line.size = 1, text.scale = 1,
  empty.intersections = "on",
   queries = list(
     list(query = intersects, params = list("Fresh"), color = "green", active = T),
     list(query = intersects, params = list("Fixed"), color = "orange", active = T)
     )
  )
```

## Manual exploration
```{r}
genes_prot_df <- fromList_togetdata(list(
    "Fresh" = rownames(seurat_obj_list_prot$Fresh),
    "Fixed" = rownames(seurat_obj_list$Fixed)
  ))
genes_prot_df <- genes_prot_df[order(row.names(genes_prot_df)), ]
  
head(genes_prot_df)
```


# Captured genes in ANY library
```{r}
genes_dataall <- Matrix::rowMeans(seurat_obj[["RNA"]]@data)
length(genes_dataall)
print(summary(genes_dataall))

# Plot
g_numall <- genes_dataall
names(g_numall) <- NULL
plot(density(g_numall))
```

# Captures genes in ALL libraries
```{r}
# Subsetting
genes_common <- genes_lib_df %>% 
  filter((Fresh_1 == 1 & Fresh_2 == 1) & (Fixed_1 == 1 & Fixed_2 == 1))
genes_common <- rownames(genes_common)
length(genes_common)

genes_common_data <- rowMeans(seurat_obj[["RNA"]]@data[genes_common,])
print(summary(genes_common_data))

# Plot
gcommon_num <- genes_common_data
names(gcommon_num) <- NULL
plot(density(gcommon_num))
```

## By PROTOCOL
### Fixed
```{r}
# Subsetting
seurat_obj_fixed <- subset(seurat_obj, subset = sample_protocol == "Fixed")
genes_common_data_fixed <- rowMeans(seurat_obj_fixed[["RNA"]]@data[genes_common,])
length(genes_common_data_fixed)
print(summary(genes_common_data_fixed))

# Plot
gcommon_num_fix <- genes_common_data_fixed
names(gcommon_num_fix) <- NULL
plot(density(gcommon_num_fix))
```

### Fresh
```{r}
seurat_obj_fresh <- subset(seurat_obj, subset = sample_protocol == "Fresh")
genes_common_data_fresh <- rowMeans(seurat_obj_fresh[["RNA"]]@data[genes_common,])
length(genes_common_data_fresh)
print(summary(genes_common_data_fresh))

#Plot
gcommon_num_fresh <- genes_common_data_fresh
names(gcommon_num_fresh) <- NULL
plot(density(gcommon_num_fresh))
```


# Captured genes by PROTOCOL (in BOTH replicates)
## Fixed
```{r}
# Subsetting
genes_fixncut_unique <-  genes_lib_df %>% 
  filter(Fresh_1 == 0 & Fresh_2 == 0 & (Fixed_1 == 1 & Fixed_2 == 1))
genes_fixncut_unique <- rownames(genes_fixncut_unique)
length(genes_fixncut_unique)

genes_fixncut_data <- rowMeans(seurat_obj_fixed[["RNA"]]@data[genes_fixncut_unique,])
print(summary(genes_fixncut_data))

#Plot
gfixed_num <- genes_fixncut_data
names(gfixed_num) <- NULL
plot(density(gfixed_num))
```

## Fresh
```{r}
# Subsetting
genes_fresh_unique <- genes_lib_df %>% 
  filter((Fresh_1 == 1 & Fresh_2 == 1) & Fixed_1 == 0 & Fixed_2 == 0)
genes_fresh_unique <- rownames(genes_fresh_unique)
length(genes_fresh_unique)

genes_fresh_data <- rowMeans(seurat_obj_fresh[["RNA"]]@data[genes_fresh_unique,])
print(summary(genes_fresh_data))

# Plots
gfresh_num <- genes_fresh_data
names(gfresh_num) <- NULL
plot(density(gfresh_num))
```

# Session Info
```{r}
sessionInfo()
```