# FIXnCUT_benchmarking

This is the GitHub repository for the **FixNCut: Single-cell genomics through reversible tissue fixation and dissociation** manuscript.

*This manuscript is currently under review process.*


## Abstract

The use of single-cell technologies for clinical applications requires disconnecting sampling from downstream processing steps. Early sample preservation can further increase robustness and reproducibility by avoiding artifacts introduced during specimen handling. We present FixNCut, a methodology for the reversible fixation of tissue followed by dissociation that overcomes current limitations. We applied FixNCut to human and mouse tissues to demonstrate the preservation of RNA integrity, sequencing library complexity, and cellular composition, while diminishing stress-related artifacts. Besides single-cell RNA sequencing, FixNCut is compatible with multiple single-cell and spatial technologies, making it a versatile tool for robust and flexible study designs.


## Code implementation

The repository is organized into the following folder tree, which contains all the necessary files and scripts to perform the detailed tasks and reproduce all our results.

* **01_cellranger_mapping** -->
* **02_QC** -->
* **03_clustering_annotation** -->
* **04_GEX_analysis** -->
* **bin** -->


## Data accessibility

* The complete raw data (FASTQ files) generated in this study have been submitted to the NCBI Gene Expression Omnibus (GEO) under accession number XXXXX.
* The count matrices and metadata are deposited at [Zenodo](https://zenodo.org/record/7837624#.ZD5RuI5BxH4). 


## Code accessibility

You can easily download a copy of all the files contained in this repository by:

* Cloning the git repository using the following command in the terminal:

`git clone https://github.com/LJimenezGracia/FIXnCUT_benchmarking.git`

* Downloading a .ZIP archive [HERE](https://github.com/LJimenezGracia/FIXnCUT_benchmarking/archive/refs/heads/main.zip).
