---
title: "Pseudobulk correlation"
author: "Laura Jiménez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
Following with the results obtained in the previous notebook, we will now perform other pre-processing steps, such as data normalization, feature selection, scaling, dimensionality reduction, and data visualization prior to batch-effect correction (data integration). To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746) and adapt some workflows from [Satija Lab](https://satijalab.org/seurat/vignettes.html).

## Objective
In this Rmarkdown document, we are going to perform the previous pre-processing steps.

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(gt)

library(edgeR)
library(SingleCellExperiment)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_r_objects <- here::here("02_QC/results/R_objects")

# Functions
source(here::here("bin/utils.R"))

# Parameters
confounder_variables <- c("library_name", "sample_protocol", "fixation_time")
confounder_names <- c("Library", "Sample Protocol", "Fixation time")

metadata_variables_confounder <- c("nCount_RNA", "nFeature_RNA", "gem_id", 
                                   "pct_mt", "pct_rb", "Phase",
                                   "library_name", "sample_protocol", "fixation_time")

```

## Load metadata
The data used in this Rmarkdown document comes from **SCGTEST_34** dataset.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)

# Merging metadata tables
metadata <- metadata %>% 
  filter(subproject == "SCGTEST_34" & tissue == "PBMC")

print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))
```


# Explore QC pre-filtering
```{r}
# Load Seurat object
seurat_obj <- readRDS(paste0(path_r_objects, "/scgtest34_h_PBMC_healthy_merged.rds"))
seurat_obj
```


Here, we show an overview of the scRNA-seq data obtained after the quality control.
```{r}
table_qc_gex(seurat_obj@meta.data, subtitle = "Before cell QC filtering out")
```

```{r fig.height=5, fig.width=14}
gg_qc_by_gemid1 <- VlnPlot(seurat_obj,
                          features = c("nCount_RNA", "nFeature_RNA"),
                          group.by = "library_name",
                          split.by = "sample_protocol",
                          pt.size = 0,
                          ncol = 2
                          ) & 
  scale_y_log10() &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
                 geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .75)) #&
  #ggpubr::stat_compare_means(ref.group = "Lung_fresh", method = "wilcox.test", label = "p.format") # p.format)

gg_qc_by_gemid2 <- VlnPlot(seurat_obj,
                          features = "pct_mt",
                          group.by = "library_name",
                          split.by = "sample_protocol",
                          pt.size = 0) +
  ylim(0, 25) +
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
                 geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .75))
#+
  #ggpubr::stat_compare_means(ref.group = "Lung_fresh", method = "wilcox.test", label = "p.format") # p.format)

cowplot::plot_grid(gg_qc_by_gemid1, gg_qc_by_gemid2, ncol = 2, rel_widths = c(1.25, 1))
```


# HVG across samples
```{r}
seurat_obj <- readRDS(paste0(path_r_objects, "/scgtest34_h_PBMC_healthy_lognorm_processed.rds"))
seurat_obj
```

```{r}
# Common HVG
seurat_obj <- FindVariableFeatures(seurat_obj, nfeatures = 3000)
common_HVG <- VariableFeatures(seurat_obj)
```


```{r}
seurat_obj_list <- SplitObject(seurat_obj, split.by = "library_name")

# Fresh HVG
seurat_obj_list$CNAG81_fresh <- FindVariableFeatures(seurat_obj_list$CNAG81_fresh, nfeatures = 3000)
fresh_HVG <- VariableFeatures(seurat_obj_list$CNAG81_fresh)

# Fix&Cut HVG
seurat_obj_list$CNAG81_fixed <- FindVariableFeatures(seurat_obj_list$CNAG81_fixed, nfeatures = 3000)
fixed_HVG <- VariableFeatures(seurat_obj_list$CNAG81_fixed)
```



```{r fig.height=5, fig.width=10}
UpSetR::upset(
  UpSetR::fromList(list(
    "PBMC_all" = common_HVG,
    "PBMC_fixed_24h" = fixed_HVG,
    "PBMC_fresh" = fresh_HVG
  )),
  nsets = 6,
  order.by = "freq",
  #group.by = "sets",
  mainbar.y.label = "# Genes", sets.x.label = "# Genes / sample",
  point.size = 4, line.size = 1,
  empty.intersections = "on",
  queries = list(
     list(query = UpSetR::intersects, params = list("PBMC_all", "PBMC_fixed_24h", "PBMC_fresh"), color = "green", active = T) )
)
```


## Exclusive HVG
```{r}

```


# Pseudo-bulk gene expression profiles & correlation
https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html

```{r}
# Load Seurat object
path_r_objects <- here::here("03_clustering_annotation/results/R_objects")
seurat_obj <- readRDS(str_c(path_r_objects, "/scgtest34_h_PBMC_healthy_cells1_clustering_annotation.rds"))

seurat_obj <- subset(seurat_obj, celltypes_1 != "Low quality cells")
seurat_obj
```


```{r}
# Extract raw counts and metadata to create SingleCellExperiment object
counts <- seurat_obj@assays$RNA@counts 
metadata <- seurat_obj@meta.data

# Create single cell experiment object
sce <- SingleCellExperiment(assays = list(counts = counts), 
                           colData = metadata)
```


```{r}
# Generate sample level metadata
## Determine the number of cells per sample
table(sce$library_name)

## Turn named vector into a numeric vector of number of cells per sample
n_cells <- as.numeric(table(sce$library_name))

## Determine how to reoder the samples (rows) of the metadata to match the order of sample names in side vector
m <- match(c("CNAG81_fresh", "CNAG81_fixed"), sce$library_name)

## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
ei <- data.frame(colData(sce)[m, ], 
                  n_cells, row.names = NULL) %>% 
                select(-"library_name")
ei
```


```{r}
# Prior to performing the aggregation of cells to the sample level, 
# we want to make sure that the poor quality cells are removed if this step hasn’t already been performed.

# Perform QC if not already performed
# Calculate quality control (QC) metrics
# Get cells w/ few/many detected genes
# Remove outlier cells
## Remove lowly expressed genes which have less than 10 cells with any counts
```


```{r}
# Aggregate the counts per sample_id
# Subset metadata to only include the cluster and sample IDs to aggregate across
groups <- colData(sce)[, "library_name"]

# Aggregate across cluster-sample groups
pb <- Matrix.utils::aggregate.Matrix(t(counts(sce)), 
                       groupings = groups, fun = "sum") 

class(pb)
dim(pb)
pb[1:2, 1:5]
```

```{r}
# Data scaling + log
MM <- apply(t(pb), 2, function(x){
      if(max(x) > 0){
        x/max(x)
      }else{
        x
      }
    })

# Scale the scaling factor
MM <- 1e4 * MM

# LogTransform
MM <- log10(MM + 1)
MM[1:8, 1:2]

MM <- as.data.frame(MM)
head(MM)
```


```{r fig.width=6, fig.height=5}
# Pearson correlation between 2 variables
ggpubr::ggscatter(MM, x = "CNAG81_fresh", y = "CNAG81_fixed", 
        add = "reg.line", # Add regression line
        add.params = list(color = "blue", fill = "lightgray"),
        conf.int = TRUE, # Add confidence interval
        cor.coef = TRUE, 
        cor.method = "pearson", # Add correlation coefficient.
          xlab = "Fresh", ylab = "Fixed", 
        title = "Pseudo-bulk Pearson correlation")
```


## By hallmark
```{r}
# Getting HALLMARK genes
HALLMARK_APOPTOSIS <- msigdbr::msigdbr(species = "human", category = "H") %>%
  filter(gs_name == "HALLMARK_APOPTOSIS")
HALLMARK_APOPTOSIS <- HALLMARK_APOPTOSIS$gene_symbol

HALLMARK_HYPOXIA <- msigdbr::msigdbr(species = "human", category = "H") %>%
  filter(gs_name == "HALLMARK_HYPOXIA")
HALLMARK_HYPOXIA <- HALLMARK_HYPOXIA$gene_symbol

HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY <- msigdbr::msigdbr(species = "human", category = "H") %>%
  filter(gs_name == "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY")
HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY <- HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY$gene_symbol

HALLMARK_G2M_CHECKPOINT <- msigdbr::msigdbr(species = "human", category = "H") %>%
  filter(gs_name == "HALLMARK_G2M_CHECKPOINT")
HALLMARK_G2M_CHECKPOINT <- HALLMARK_G2M_CHECKPOINT$gene_symbol
```

### HALLMARK_APOPTOSIS
```{r fig.width=6, fig.height=5}
MM_subset <- MM %>% 
  filter(row.names(MM) %in% HALLMARK_APOPTOSIS)

# Pearson correlation between 2 variables
ggpubr::ggscatter(MM_subset, x = "CNAG81_fresh", y = "CNAG81_fixed", 
          add = "reg.line", # Add regression line
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, # Add confidence interval
          cor.coef = TRUE, 
          cor.method = "pearson", # Add correlation coefficient.
          xlab = "Fresh", ylab = "Fixed", 
          title = "Pseudo-bulk Pearson correlation _ HALLMARK_APOPTOSIS")
```


### HALLMARK_HYPOXIA
```{r fig.width=6, fig.height=5}
MM_subset <- MM %>% 
  filter(row.names(MM) %in% HALLMARK_HYPOXIA)

# Pearson correlation between 2 variables
ggpubr::ggscatter(MM_subset, x = "CNAG81_fresh", y = "CNAG81_fixed", 
          add = "reg.line", # Add regression line
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, # Add confidence interval
          cor.coef = TRUE, 
          cor.method = "pearson", # Add correlation coefficient.
          xlab = "Fresh", ylab = "Fixed", 
          title = "Pseudo-bulk Pearson correlation _ HALLMARK_HYPOXIA")
```

### HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY
```{r fig.width=6, fig.height=5}
MM_subset <- MM %>% 
  filter(row.names(MM) %in% HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY)

# Pearson correlation between 2 variables
ggpubr::ggscatter(MM_subset, x = "CNAG81_fresh", y = "CNAG81_fixed", 
          add = "reg.line", # Add regression line
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, # Add confidence interval
          cor.coef = TRUE, 
          cor.method = "pearson", # Add correlation coefficient.
          xlab = "Fresh", ylab = "Fixed", 
          title = "Pseudo-bulk Pearson correlation _ HALLMARK_ROS_PATHWAY")
```

### HALLMARK_G2M_CHECKPOINT
```{r fig.width=6, fig.height=5}
MM_subset <- MM %>% 
  filter(row.names(MM) %in% HALLMARK_G2M_CHECKPOINT)

# Pearson correlation between 2 variables
ggpubr::ggscatter(MM_subset, x = "CNAG81_fresh", y = "CNAG81_fixed", 
          add = "reg.line", # Add regression line
          add.params = list(color = "blue", fill = "lightgray"),
          conf.int = TRUE, # Add confidence interval
          cor.coef = TRUE, 
          cor.method = "pearson", # Add correlation coefficient.
          xlab = "Fresh", ylab = "Fixed", 
          title = "Pseudo-bulk Pearson correlation _ HALLMARK_G2M_CHECKPOINT")
```


# Pseudo-bulk gene expression profiles & correlation by cell-type
https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html

```{r}
# Load Seurat object
path_r_objects <- here::here("03_clustering_annotation/results/R_objects")

seurat_obj <- readRDS(str_c(path_r_objects, "/scgtest34_h_PBMC_healthy_cells1_clustering_annotation.rds"))
seurat_obj

Idents(seurat_obj) <- "library_name"
seurat_obj <- RenameIdents(seurat_obj,
             "CNAG81_fresh" = "CNAG81-fresh",
             "CNAG81_fixed" = "CNAG81-fixed"
             )
seurat_obj$library_name <- as.character(seurat_obj@active.ident)
Idents(seurat_obj) <- "celltypes_1"
```


```{r}
# Print out the table of cells in each cluster-sample group
table(seurat_obj$celltypes_1, seurat_obj$library_name)
```


```{r}
# Extract raw counts and metadata to create SingleCellExperiment object
counts <- seurat_obj@assays$RNA@counts 
metadata <- seurat_obj@meta.data

# Create single cell experiment object
sce <- SingleCellExperiment(assays = list(counts = counts), 
                           colData = metadata)
```


```{r}
# Generate sample level metadata
## Determine the number of cells per sample
table(sce$library_name)

## Turn named vector into a numeric vector of number of cells per sample
n_cells <- as.numeric(table(sce$library_name))

## Determine how to reoder the samples (rows) of the metadata to match the order of sample names in side vector
m <- match(c("CNAG81-fresh", "CNAG81-fixed"), sce$library_name)

## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
ei <- data.frame(colData(sce)[m, ], 
                  n_cells, row.names = NULL) %>% 
                select(-"library_name")
ei
```


```{r}
# Prior to performing the aggregation of cells to the sample level, 
# we want to make sure that the poor quality cells are removed if this step hasn’t already been performed.

# Perform QC if not already performed
# Calculate quality control (QC) metrics
# Get cells w/ few/many detected genes
# Remove outlier cells
## Remove lowly expressed genes which have less than 10 cells with any counts
```


```{r}
# Aggregate the counts per sample_id
# Subset metadata to only include the cluster and sample IDs to aggregate across
groups <- colData(sce)[, c("celltypes_1", "library_name")]

# Aggregate across cluster-sample groups
pb <- Matrix.utils::aggregate.Matrix(t(counts(sce)), 
                       groupings = groups, fun = "sum") 

class(pb)
dim(pb)
pb[1:5, 1:8]
```


```{r}
# Not every cluster is present in all samples; create a vector that represents how to split samples
splitf <- sapply(stringr::str_split(rownames(pb), pattern = "_",  n = 2), `[`, 1)
head(splitf)
```


```{r}
# Turn into a list and split the list into components for each cluster and transform, 
# so rows are genes and columns are samples and make rownames as the sample IDs
pb <- split.data.frame(pb, factor(splitf)) %>%
        lapply(function(u) trqwe::set_colnames(t(u), stringr::str_extract(rownames(u), "_(.*)")))

# Explore the different components of list
str(pb)
```

```{r}
# Print out the table of cells in each cluster-sample group
table(sce$celltypes_1, sce$library_name)
```



```{r fig.width=12, fig.height=10}
for (cluster in sort(unique(seurat_obj$celltypes_1))) {
  
  if (sum(seurat_obj$celltypes_1 == cluster) > 100 &&
      all(as.numeric(table(seurat_obj$library_name[seurat_obj$celltypes_1 == cluster])) > 25) ) {
    
    print(cluster)
    pb_sub <- t(pb[[cluster]])

    # Data scaling + log
    MM <- apply(t(pb_sub), 2, function(x){
          if(max(x) > 0){
            x/max(x)
          }else{
            x
          }
        })
    
    # Scale the scaling factor
    MM <- 1e4 * MM
    
    # LogTransform
    MM <- log10(MM + 1)

    MM <- as.data.frame(MM)
    head(MM)
    
    # Pearson correlation between 2 variables
    gg <- ggpubr::ggscatter(MM, x = "_CNAG81-fresh", y = "_CNAG81-fixed", 
                  add = "reg.line", # Add regression line
                  add.params = list(color = "blue", fill = "lightgray"),
                  conf.int = TRUE, # Add confidence interval
                  cor.coef = TRUE, 
                  cor.method = "pearson", # Add correlation coefficient.
                  xlab = "Fresh", ylab = "Fixed", 
                  title = paste0(cluster, " -bulk Pearson correlation"))
    print(gg)

  }
}
```

## Downsampling by min # cells
```{r}
pseudobulk_cond <- function(seurat_obj) {
  # Extract raw counts and metadata to create SingleCellExperiment object
  counts <- seurat_obj@assays$RNA@counts 
  metadata <- seurat_obj@meta.data
  
  # Create single cell experiment object
  sce <- SingleCellExperiment(assays = list(counts = counts), colData = metadata)
  
  ## Turn named vector into a numeric vector of number of cells per sample
  n_cells <- as.numeric(table(sce$library_name))
  
  ## Determine how to reoder the samples (rows) of the metadata to match the order of sample names in side vector
  m <- match(c("CNAG81-fresh", "CNAG81-fixed"), sce$library_name)
  
  ## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
  #ei <- data.frame(colData(sce)[m, ], n_cells, row.names = NULL) %>% 
  #  select(-"library_name")
  
  # Aggregate the counts per sample_id
  # Subset metadata to only include the cluster and sample IDs to aggregate across
  groups <- colData(sce)[, "library_name"]
  
  # Aggregate across cluster-sample groups
  pb <- Matrix.utils::aggregate.Matrix(t(counts(sce)), groupings = groups, fun = "sum") 
  
  # Data scaling + log
  MM <- apply(t(pb), 2, function(x){ if(max(x) > 0) { x/max(x) } else { x }})
  
  # Scale the scaling factor
  MM <- 1e4 * MM
  # LogTransform
  MM <- log10(MM + 1)
  MM <- as.data.frame(MM)
  
  return(MM)
}
```


```{r}
# Print out the table of cells in each cluster-sample group
table(seurat_obj$celltypes_1, seurat_obj$library_name)
```


```{r fig.width=12, fig.height=10}
set.seed(0)

for (cluster in sort(unique(seurat_obj$celltypes_1))) {
  print(cluster)
  if (sum(seurat_obj$celltypes_1 == cluster) > 100 &&
      all(as.numeric(table(seurat_obj$library_name[seurat_obj$celltypes_1 == cluster])) > 20) ) {
    
    seurat_obj_c <- seurat_obj[, seurat_obj$celltypes_1 == cluster]
    seurat_obj_c$celltypes_1 <- as.factor(as.character(seurat_obj_c$celltypes_1))
    
    Idents(seurat_obj_c) <- "library_name"
    min_cells <- min(as.numeric(table(seurat_obj_c$sample_protocol)))
    
    seurat_obj_c <- subset(seurat_obj_c, downsample = min_cells)
    
    #keep_genes <- rownames(seurat_obj_c)[Matrix::rowSums(seurat_obj_c[["RNA"]]@counts > 0) > 0]
    #seurat_obj_c <- subset(seurat_obj_c, features = keep_genes)
    
    Idents(seurat_obj_c) <- "library_name"

    
    MM <- pseudobulk_cond(seurat_obj_c)
    
    # Pearson correlation between 2 variables
    gg <- ggpubr::ggscatter(MM, x = "CNAG81-fresh", y = "CNAG81-fixed", 
              add = "reg.line", # Add regression line
              add.params = list(color = "blue", fill = "lightgray"),
              conf.int = TRUE, # Add confidence interval
              cor.coef = TRUE, 
              cor.method = "pearson", # Add correlation coefficient.
              xlab = "Fresh", ylab = "Fixed", 
              title = paste0(cluster, " -bulk Pearson correlation"))

    print(gg)
  }
}
```


# Session Info
```{r}
sessionInfo()
```