---
title: "GEX Quality control: MouseLung_fixed_perfused"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
As shown in the previous notebook, there are no major differences among all libraries processed and sequenced at CNAG in terms of sequencing depth, number of recovered cells, mean reads per cell, median genes per cell, among others. For this reason, we will merge all libraries together and we will perform a single quality control (QC) including all libraries. Therefore, unique QC thresholds will be determined and applied to all libraries.

## Objective
In this Rmarkdown document, we are going to perform a quality control (QC) in order to filter out low-quality cells and genes. To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746).

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(ggpubr)
library(gt)
library(grid)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_cellranger_data <- here::here(paste0("02_QC/1_downsampling_reads"))
#path_scrublet_data <- here::here("02_QC/2_doublet_detection/results/tables")
path_r_objects <- here::here("02_QC/results/R_objects")

# Functions
source(here::here("bin/utils.R"))

# Thresholds
min_libsize <- 400
max_libsize <- 50000

min_ngenes <- 200
max_ngenes <- 8000

max_pctmt <- 20

min_cells <- 10
```


## Load data
The data used in this Rmarkdown document comes from **SCGTEST_33** dataset, filtered matrices were processed with `cellranger v6.1.1`, and the doublet predictions were obtained using `scrublet`.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)
subproject_folder_n <- "MouseLung_fixed_perfused"

# Merging metadata tables
metadata <- metadata %>% 
  filter(subproject == "SCGTEST_33"
         & tissue == "Lung"
         & ( sample_protocol == "Fresh" | sample_protocol == "Fix&Cut_24h"))

# Modify metadata
metadata$library_name[metadata$library_name == "Lung_fresh"] <- "Fresh"
metadata$library_name[metadata$library_name == "Lung_fixed_24h"] <- "Fixed"

metadata$sample_protocol <- metadata$library_name

print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))
```


```{r}
# Metadata
aggr_csv_path <- paste0(path_cellranger_data, "/data/", subproject_folder_n, "_aggr.csv")
metadata_aggr <- read.csv(aggr_csv_path)

# Modify metadata
metadata_aggr$library_name[metadata_aggr$library_name == "Lung_fresh"] <- "Fresh"
metadata_aggr$library_name[metadata_aggr$library_name == "Lung_fixed_24h"] <- "Fixed"

for (index in rownames(metadata_aggr)) {
  print(index)
  lib_name <- metadata_aggr[index, ]$library_name
  metadata_aggr[index, ]$gem_id <- metadata$gem_id[metadata$library_name == lib_name]
}

# Create Seurat object
## Counts data
filtered_matrix_path <- paste0(path_cellranger_data, "/", subproject_folder_n, "/outs/count/filtered_feature_bc_matrix")
expression_matrix <- Read10X(data.dir = filtered_matrix_path)
seurat_obj <- CreateSeuratObject(counts = expression_matrix, project = subproject_folder_n)

# Rename cellnames / columns
gem_id_list <- purrr::map(colnames(seurat_obj), function(colname){
  x <- str_split(colname, "-")[[1]]
  gem_id <- metadata_aggr[x[2], ]$sample_id
  gem_id
  })
gem_id_list <- unlist(gem_id_list)
seurat_obj <- RenameCells(seurat_obj, add.cell.id = gem_id_list)
head(colnames(seurat_obj))

# Add orig.init and gem_id metadata
seurat_obj$orig.ident <- gem_id_list
seurat_obj$gem_id <- gem_id_list
```


```{r}
seurat_obj <- RenameCells(seurat_obj,
                          new.names = str_replace(colnames(seurat_obj), pattern = regex("-[0-9]"), replacement = "-1"))
head(colnames(seurat_obj))
tail(colnames(seurat_obj))
```


## Add library metadata
Now, we will add metadata_all information of the different samples.
```{r}
# Create dataframe with seurat metadata_all (by barcodes) and sample metadata_all
seurat_obj_metadata <- left_join(seurat_obj@meta.data, metadata, by = "gem_id")

# Add only relevant metadata_all variables & convert into factors
seurat_obj$library_name <- as.factor(seurat_obj_metadata$library_name)
seurat_obj$donor_id <- as.factor(seurat_obj_metadata$organism)
seurat_obj$tissue <- as.factor(seurat_obj_metadata$tissue)
seurat_obj$disease <- as.factor(seurat_obj_metadata$disease)
seurat_obj$sample_protocol <- as.factor(seurat_obj_metadata$sample_protocol)
```


## Add scrublet prediction
```{r}
seurat_obj_list <- SplitObject(seurat_obj, split.by = "gem_id")

# Add doublet predictions
gem_ids <- unique(gem_id_list)
scrublet_files_list <- purrr::map(unique(gem_ids), function(gem_id) {
  subproject <- unique(metadata$subproject[metadata$gem_id == gem_id])
  path_scrublet <- here::here(paste0(
  "02_QC/0_doublet_detection/results/", subproject, "__", gem_id, "_doublet_prediction_scrublet.csv"))
}
)

scrublet_dfs <- lapply(scrublet_files_list, function(scrublet_file_path){
  read.csv(scrublet_file_path)
})
names(scrublet_dfs) <- gem_ids

# Rename barcodes with gem_id
for (gem_id in names(scrublet_dfs)) {
  scrublet_dfs[[gem_id]]$barcodes <- paste0(gem_id, "_", scrublet_dfs[[gem_id]]$barcodes)
}

# Integrate scrublet doublet predictions to seurat metadata
seurat_obj_list <- purrr::map2(seurat_obj_list, 
                                          names(seurat_obj_list), 
                                          function(seurat, gem_id) {
                                            
  if (all(colnames(seurat) == scrublet_dfs[[gem_id]]$barcodes)) {
    seurat$scrublet_doublet_scores <- scrublet_dfs[[gem_id]]$scrublet_doublet_scores
    seurat$scrublet_predicted_doublet <- scrublet_dfs[[gem_id]]$scrublet_predicted_doublet
    seurat
  } 
})

seurat_obj_list
```


## Merge Seurat list
```{r}
seurat_obj <- merge(seurat_obj_list[[1]], 
                    y = seurat_obj_list[2:length(seurat_obj_list)])

# Remove Seurat object list
rm(seurat_obj_list)
```


# GEX quality control
First, we want to have an idea about the libraries quality. For this reason, we will plot the main QC variables and check how they distribute across libraries. This will help us interpret the results downstream, as we will know which are the libraries with lower quality.

It is important to keep in mind that every library contains cells from a single mice, and we have two replicates for each sample_protocol.
```{r fig.width=14, fig.height=10}
# Plot library complexity
gg_libsize <- seurat_obj@meta.data %>%
  #dplyr::filter(HTO_classification.global == "Singlet") %>%
    gg_gex_horizontal_boxplot(
    categorical_var = "library_name",
    continuous_var = "nCount_RNA",
    fill = "sample_protocol",
    title = "Library size",
    ylab = "Total UMIs",
    decreasing = TRUE
  ) + scale_y_log10() +
  theme(legend.position = "none")

# Plot library complexity
gg_libcomplex <- seurat_obj@meta.data %>%
  #dplyr::filter(HTO_classification.global == "Singlet") %>%
    gg_gex_horizontal_boxplot(
    categorical_var = "library_name",
    continuous_var = "nFeature_RNA",
    fill = "sample_protocol",
    title = "Library complexity",
    ylab = "Number of Detected Genes",
    decreasing = TRUE
  ) + scale_y_log10() +
  theme(legend.position = "none")

# Calculate percentage mitocondrial genes
seurat_obj$pct_mt <- PercentageFeatureSet(
  seurat_obj,
  pattern = "^mt-"
)

# Plot MT expression
gg_mt <- seurat_obj@meta.data %>%
  #dplyr::filter(HTO_classification.global == "Singlet") %>%
    gg_gex_horizontal_boxplot(
    categorical_var = "library_name",
    continuous_var = "pct_mt",
    fill = "sample_protocol",
    title = "Mitochondrial Expression",
    ylab = "Fraction of Mitochondrial Genes (%)",
    decreasing = TRUE
  ) +
  theme(legend.position = "none")

# Calculate percentage ribosomal genes
seurat_obj$pct_rb <- PercentageFeatureSet(
  seurat_obj,
  pattern = "^Rp[sl]")

# Plot RB expression
gg_rb <- seurat_obj@meta.data %>%
  #dplyr::filter(HTO_classification.global == "Singlet") %>%
    gg_gex_horizontal_boxplot(
    categorical_var = "library_name",
    continuous_var = "pct_rb",
    fill = "sample_protocol",
    title = "Ribosomal Expression",
    ylab = "Fraction of Ribosomal Genes (%)",
    decreasing = TRUE
  ) +
  theme(legend.position = "none")


cowplot::plot_grid(gg_libsize, gg_libcomplex, gg_mt, gg_rb, ncol = 2)
```

## Save merged data
```{r}
# Save Merged Seurat object
saveRDS(seurat_obj, str_c(path_r_objects, "/FIXnCUT_MouseLung_fixed_perfused_merged.rds"))
#seurat_obj <- readRDS(str_c(path_r_objects, "/FIXnCUT_MouseLung_fixed_perfused_merged.rds"))
```


# Cell QC
At this point, and before analyzing gene expression data, we must ensure that we will only work with single and viable cells. This means we need to detect all outlier cell barcodes, and filter them because they can correspond to empty droplets, dying or lysed cells. For this reason, we will be visualize and examine the distribution of three QC covariates, which are indeed hallmarks of dying cells:
(1) Low library size  (Total UMIs)
(2) Low library complexity (Number of Detected Genes), and 
(3) High fraction of counts from mitochondrial genes (cytosolic mRNA leaks out of the cell).
Finally, we would also explore the fraction of counts from ribosomal genes, but no filtering would be applied to them.

## Library size
```{r fig.height=5, fig.width=14}
ggdens_libsize <- seurat_obj@meta.data %>%
  ggplot(aes(x = nCount_RNA, fill = sample_protocol, color = sample_protocol)) +
  #facet_wrap(~replicate, nrow = 2) +
    geom_density(alpha = 0.2) + 
    scale_x_log10() +
    labs(x = "Library Size (Total UMI)",
         y = "Density") +
    theme_pubr() +
    geom_vline(xintercept = min_libsize, linetype = "dashed", color = "red") +
    geom_vline(xintercept = max_libsize, linetype = "dashed", color = "red") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          legend.position = "bottom") +
    guides(color = FALSE, fill = guide_legend(title = "sample_protocol", ncol=3))

gghist_libsize <- seurat_obj@meta.data %>%
  ggplot(aes(nCount_RNA)) +
    geom_histogram(bins = 100) +
    labs(x = "Library Size (Total UMI)",
         y = "Frequency") +
    theme_pubr() +
    geom_vline(xintercept = min_libsize, linetype = "dashed", color = "red") +
    geom_vline(xintercept = max_libsize, linetype = "dashed", color = "red")

cowplot::plot_grid(ggdens_libsize, gghist_libsize, ncol = 2)
```

The count depth (UMIs per cell) should generally be above 500 when 3' gene expression technologies are applied, that is the low end of what we expect. However, for the 5' gene expression experiments we expected to get lower quality results, this means fewer number of UMIs per cell.

## Library complexity
```{r fig.height=5, fig.width=14}
ggdens_ngenes <- seurat_obj@meta.data %>%
  ggplot(aes(x = nFeature_RNA, fill = sample_protocol, color = sample_protocol)) +
    geom_density(alpha = 0.2) + 
  #facet_wrap(~replicate, nrow = 2) +
    scale_x_log10() +
    labs(x = "Library complexity (Number of Detected Genes)",
         y = "Density") +
    theme_pubr() +
    geom_vline(xintercept = min_ngenes, linetype = "dashed", color = "red") +
    geom_vline(xintercept = max_ngenes, linetype = "dashed", color = "red") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          legend.position = "bottom") +
    guides(color = FALSE, fill = guide_legend(title = "sample_protocol"))


gghist_ngenes <- seurat_obj@meta.data %>%
  ggplot(aes(nFeature_RNA)) +
    geom_histogram(bins = 100) +
    labs(x = "Library complexity (Number of Detected Genes)",
         y = "Frequency") +
    theme_pubr() +
    geom_vline(xintercept = min_ngenes, linetype = "dashed", color = "red") +
    geom_vline(xintercept = max_ngenes, linetype = "dashed", color = "red")

cowplot::plot_grid(ggdens_ngenes, gghist_ngenes, ncol = 2)
```

Ideally, the density distribution of high-quality data should contain a single large peak. If we see a small shoulder to the right of the major peak (not present in our data), or a bimodal distribution of the cells, that can indicate the failure of a set of cells, a group of cells that are biologically different types (i.e. quiescent cells), and/or presence of cell types much smaller than the others.


## Fraction of mitochondrial expression
```{r fig.height=5, fig.width=14}
ggdens_pctmt <- seurat_obj@meta.data %>%
  ggplot(aes(x = pct_mt, fill = sample_protocol, color = sample_protocol)) +
    geom_density(alpha = 0.2) + 
  #facet_wrap(~replicate, nrow = 2) +
    labs(x = "% Mitochondrial Expression",
         y = "Density") +
    theme_pubr() +
    scale_x_continuous(limits = c(0, 50)) +
    geom_vline(xintercept = max_pctmt, linetype = "dashed", color = "red") +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          legend.position = "bottom") +
    guides(color = FALSE, fill = guide_legend(title = "sample_protocol"))


gghist_pctmt <- seurat_obj@meta.data %>%
  ggplot(aes(pct_mt)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(limits = c(0, 100)) +
    labs(x = "% Mitochondrial Expression",
         y = "Frequency") +
    theme_pubr() +
    geom_vline(xintercept = max_pctmt, linetype = "dashed", color = "red")

cowplot::plot_grid(ggdens_pctmt, gghist_pctmt, ncol = 2)
```

High mitochondrial read fractions could be indicative of damaged/dying cells whose cytoplasmic mRNA has leaked out through a broken membrane, and thus, only mRNA located in the mitochondria is still conserved.

## Fraction of ribosomal expression
```{r fig.height=5, fig.width=14}
ggdens_pctrb <- seurat_obj@meta.data %>%
  ggplot(aes(x = pct_rb, fill = sample_protocol, color = sample_protocol)) +
    geom_density(alpha = 0.2) + 
  #facet_wrap(~replicate, nrow = 2) +
    labs(x = "% Ribosomal Expression",
         y = "Density") +
    theme_pubr() +
    theme(axis.title = element_text(size = 12),
          axis.text = element_text(size = 10),
          legend.position = "bottom") +
    guides(color = FALSE, fill = guide_legend(title = "sample_protocol"))

gghist_pctrb <- seurat_obj@meta.data %>%
  ggplot(aes(pct_rb)) +
    geom_histogram(bins = 100) +
    scale_x_continuous(limits = c(0, 100)) +
    labs(x = "% Ribosomal Expression",
         y = "Frequency") +
    theme_pubr()

cowplot::plot_grid(ggdens_pctrb, gghist_pctrb, ncol = 2)
```


## Multiplets
```{r fig.width=14, fig.height=5}
VlnPlot(seurat_obj,
        features = c("nCount_RNA", "nFeature_RNA", "scrublet_doublet_scores"),
        split.by = "scrublet_predicted_doublet",
        group.by = "library_name")
```


```{r fig.width=12, fig.height=6}
gg_cell_num <- dittoSeq::dittoBarPlot(seurat_obj, 
                            "scrublet_predicted_doublet",
                            group.by = "library_name",
                            scale = "count",
                            retain.factor.levels = TRUE,
                            legend.show = FALSE)

gg_cell_prop <- dittoSeq::dittoBarPlot(seurat_obj, 
                             "scrublet_predicted_doublet",
                             group.by = "library_name",
                             scale = "percent",
                            retain.factor.levels = TRUE)

gg_cell_num + gg_cell_prop
```



## Joint QC metrics
After examining the previous results, we can see that the distribution of the three QC covariates is really similar among different libraries. For this reason, we will determine common QC thresholds for all of them.

It is important not to consider the QC covariates independently as this can lead to the misinterpretation of results (e.g. cells with low counts could correspond to quiescent cells, whereas cells with high counts may be bigger, metabolically active cells might also have higher mitochondrial expression). For this reason, we will jointly explore how these three variables covary in order to set univariate thresholds.

```{r fig.height=10, fig.width=14}
# Library size / count depth
ggdens_libsize <- seurat_obj@meta.data %>% 
    ggplot(aes(x = nCount_RNA)) + 
    geom_density() + 
    scale_x_log10() +
  labs(x = "Library size (total UMI)",
       y = "Density") +
  theme_pubr() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

ggdens_libsize_cutoff <- ggdens_libsize +
  geom_vline(xintercept = min_libsize, color = "red") +
  geom_vline(xintercept = max_libsize, color = "red")


# Library complexity
ggdens_ngenes <- seurat_obj@meta.data %>% 
    ggplot(aes(x = nFeature_RNA)) + 
    geom_density() + 
    scale_x_log10() +
  labs(x = "Number of Genes",
       y = "Density") +
  theme_pubr() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

ggdens_ngenes_cutoff <- ggdens_ngenes +
  geom_vline(xintercept = min_ngenes, color = "red") +
  geom_vline(xintercept = max_ngenes, color = "red")


# Library complexity vs Library size by %MT
gg_ngenes_vs_libsize_by_pctmt <- seurat_obj@meta.data %>%
  ggplot(aes(nCount_RNA, nFeature_RNA, color = pct_mt)) +
  geom_point(size = 0.15) +
  scale_color_viridis_c() +
  theme_classic() +
  labs(x = "Library size (total UMI)",
       y = "Number of Genes",
       color = "% Mitochondrial Expression") +
  theme(axis.title = element_text(size = 12),
      axis.text = element_text(size = 10),
      legend.title = element_text(angle = 90),
      )
                             
gg_ngenes_vs_libsize_by_pctmt_cutoff <- gg_ngenes_vs_libsize_by_pctmt +
  geom_vline(xintercept = min_libsize, color = "red") +
  geom_hline(yintercept = min_ngenes, color = "red") +
  geom_vline(xintercept = max_libsize, color = "red") +
  geom_hline(yintercept = max_ngenes, color = "red")
  
  
# %MT vs Library size
gg_pctmt_vs_libsize <- FeatureScatter(
  seurat_obj,
  feature1 = "nCount_RNA",
  feature2 = "pct_mt",
  pt.size = 0.15,
  cols = rep("black", length(levels(Idents(seurat_obj))))
)

gg_pctmt_vs_libsize <- gg_pctmt_vs_libsize +
  ylim(0, 100) +
  labs(x = "Library size (total UMI)",
       y = "% Mitochondrial Expression") +
  theme(legend.position = "none",
        plot.title = element_blank())

gg_pctmt_vs_libsize_cutoff <- gg_pctmt_vs_libsize +
  geom_hline(yintercept = max_pctmt, color = "red") +
  geom_vline(xintercept = min_libsize, color = "red") +
  geom_vline(xintercept = max_libsize, color = "red")

cowplot::plot_grid(plotlist = list(ggdens_libsize_cutoff,
                                   ggdens_ngenes_cutoff,
                                   gg_ngenes_vs_libsize_by_pctmt_cutoff,
                                   gg_pctmt_vs_libsize_cutoff),
                   ncol = 2)
```

In the bottom left graph, we have plotted the number of genes vs the number of counts colored by the fraction of mitochondrial reads. As it can be observed, mitochondrial read fractions are only high in particularly low count cells with few detected genes, which will be filtered out by the minimum count depth and gene number thresholds. Good cells will generally exhibit both higher number of genes and counts per cell.

```{r}
# Metadata before cell filtering
metadata_before_qc <- seurat_obj@meta.data
```


## Filter out low quality cells

Different threshold values have been selected according to the previous results. On one side, cell barcodes with a low count depth (< `r min_libsize`), few detected genes (< `r min_ngenes`), and a high fraction of mitochondrial counts (> `r max_pctmt`%) are indicative of lysed cells (cytoplasmic mRNA was lost through the broken membrane, and only mitochondria mRNA was conserved). Therefore, we will proceed to filter them out.

```{r}
# Subset low quality cell barcodes
low_quality_subset <-
  seurat_obj$nCount_RNA < min_libsize |
  seurat_obj$nFeature_RNA < min_ngenes |
  seurat_obj$pct_mt > max_pctmt
table(low_quality_subset)
```


```{r}
# Filtering out low quality cell barcodes
seurat_obj$keep_cells <- !low_quality_subset
Idents(seurat_obj) <- "keep_cells"
seurat_obj <- subset(seurat_obj, idents = TRUE)
seurat_obj$keep_cells <- NULL
```


## Filter out potential doublets

On the other, cells with unexpectedly high library size / count depth (> `r max_libsize`) and/or large number of detected genes (> `r max_ngenes`) may represent doublets. To avoid having doublets in our dataset, we will remove these few outlier cells.

```{r}
# Subset high library size / complexity quality cells
high_lib_size_complexity <-
  seurat_obj$nCount_RNA > max_libsize |
  seurat_obj$nFeature_RNA > max_ngenes

table(high_lib_size_complexity)
```

```{r}
# Filtering out high library size / complexity quality cells
seurat_obj$keep_cells <- !high_lib_size_complexity
Idents(seurat_obj) <- "keep_cells"
seurat_obj <- subset(seurat_obj, idents = TRUE)
seurat_obj$keep_cells <- NULL
```


```{r}
# Metadata after filtering cells
metadata_after_qc <- seurat_obj@meta.data
```


## QC summary table

```{r}
table_qc_gex(metadata_before_qc, subtitle = "Before cell QC filtering out")

table_qc_gex(metadata_after_qc, subtitle = "After cell QC filtering out")
```


# Genes QC

After ensuring the integrity of cells, we should also performed QC on the genes. snRNA-seq experiments usually include counts for a large number of genes (> 20000). However, some these genes have zero counts and can dramatically reduce the expression of a certain cell, and other genes are only expressed in a few cells and do not inform of the cellular heterogeneity, but might correspond to rare cell types lowly expressed. A possible approach to reduce the number of genes is to filter out gens expressed in few cells than a specific threshold value. According to Luecken MD et al. "A guideline to setting this threshold is to use the minimum cell cluster size that is of interest and leaving some leeway for dropout effects. The choice of threshold should scale with the number of cells in the dataset and the intended downstream analysis."

In this project, since we aim to detect specific immune cell types that might be exclusive from the Synovial Fluid or from patients, we will be quite permissive and set a threshold value to retain all genes that are expressed in at least `r min_cells` cells. In the following plot it is possible to see distribution of gene expression across cell barcodes, including the filtering threshold, as well the highly expressed genes among cells.

```{r fig.width=14, fig.height=5}
# Number of cells expressing each gene, excluding 0 counts
n_cells <- Matrix::rowSums(seurat_obj[["RNA"]]@counts > 0)

gg_ngenes_vs_cells <- n_cells %>%
  as.data.frame() %>%
  ggplot(aes(n_cells)) +
    geom_histogram(bins = 100, alpha = 0.75) +
    scale_x_log10() +
    labs(x = "Number of expressing cells",
         y = "Genes Frequency") +
    theme_pubr()

gg_ngenes_vs_cells_cutoff <- gg_ngenes_vs_cells +
  geom_vline(xintercept = min_cells, color = "red")


# Most common expressed genes
top_40_genes <- sort(n_cells, decreasing = TRUE)[1:40]
top_40_genes_df <- data.frame(
  gene = names(top_40_genes),
  n_cells = top_40_genes
)

gg_top40genes <- top_40_genes_df %>%
  ggplot(aes(fct_reorder(gene, n_cells), n_cells)) +
    geom_point() +
    labs(x = "Genes",
         y = "Number of expressing cells") +
    coord_flip() +
    theme_bw()

cowplot::plot_grid(plotlist = list(gg_ngenes_vs_cells_cutoff,
                                   gg_top40genes), ncol = 2)
```

## Filter out genes expressed in few cells
```{r}
# Subset genes highly expressed
keep_genes <- rownames(seurat_obj)[n_cells > min_cells]
table(n_cells > min_cells)
```


```{r}
# Filtering out genes expressed in few cells
seurat_obj <- subset(seurat_obj, features = keep_genes)
seurat_obj
```

# Save filtered data
```{r}
# Save Merged Seurat object
saveRDS(seurat_obj, str_c(path_r_objects, "/FIXnCUT_MouseLung_fixed_perfused_filtered.rds"))
#seurat_obj <- readRDS(str_c(path_r_objects, "/FIXnCUT_MouseLung_fixed_perfused_filtered.rds"))
```


# Session Info
```{r}
sessionInfo()
```