---
title: "scRNAseq finding stress signature"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
Following with the results obtained in the previous notebook, we will now perform other pre-processing steps, such as data normalization, feature selection, scaling, dimensionality reduction, and data visualization prior to batch-effect correction (data integration). To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746) and adapt some workflows from [Satija Lab](https://satijalab.org/seurat/vignettes.html).

## Objective
In this Rmarkdown document, we are going to perform the previous pre-processing steps.

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(svglite)
library(Seurat)
library(UCell)
library(Nebulosa)
library(viridis)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_r_objects <- here::here("03_clustering_annotation/results/R_objects")
path_r_figs_out <- here::here("04_GEX_analysis/results/figs")

# Functions
source(here::here("bin/utils.R"))

# Color palette
palette_color_factor <- c("Fresh" = "#66C2A5", 
                          "Fixed" ="#FC8D62")
```

## Load data
```{r}
# Load Seurat object
seurat_obj <- readRDS(str_c(path_r_objects, "/FIXnCUT_HumanPBMC_clustering_annotation_cleaned.rds"))
seurat_obj$condition <- factor(x = seurat_obj$condition,
                                    levels = c("Fresh", "Fixed"))
seurat_obj$celltypes <- factor(x = seurat_obj$celltypes,
                                    levels = c("Naive B cells", "Memory B cells", "Naive CD4+ T cells", "Transitional Memory CD4+ T cells",
                         "Central Memory CD4+ T cells", "Effector Memory CD4+ T cells", "Tregs", 
                         "Naive CD8+ T cells", "Cytotoxic CD8+ T cells", "gdT cells", "MAIT", "NK cells",
                         "Classical Monocytes", "Intermediate Monocytes", "Non-classical Monocytes",
                         "DC1/DC2","pDCs",
                         "Progenitor cells","Cycling cells"))
seurat_obj
```


# Housekeeping markers

Housekeeping genes used from: 
https://www.cell.com/trends/genetics/fulltext/S0168-9525(13)00089-9?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0168952513000899%3Fshowall%3Dtrue

```{r fig.width=20, fig.height=8}
housekeeping_genes <- c("ACTB", "C1orf43", "CHMP2A", "EMC7", "GPI", "PSMB2", "PSMB4", "RAB7A", "REEP5", "SNRPD3", "VCP", "VPS29")

VlnPlot(
  seurat_obj,
  features = housekeeping_genes,
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 6) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```


```{r fig.width=20, fig.height=20}
VlnPlot(
  seurat_obj,
  features = housekeeping_genes,
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 4) &
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1),
        axis.text = element_text(size=12),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```


# Low quality markers

```{r fig.width=15, fig.height=4}
lowquality_genes <- c("HBB", "HBA1", "HBA2", "HBD")

VlnPlot(
  seurat_obj,
  features = lowquality_genes,
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 4) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```

```{r fig.width=20, fig.height=5}
VlnPlot(
  seurat_obj,
  features = lowquality_genes,
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0.01, 
  ncol = 4) &
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1),
        axis.text = element_text(size=12),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```


# Heat shock protein (HSP) markers
```{r fig.width=20, fig.height=8}
HSP_genes <- c("HSP90AA1", "HSP90AB1", "HSPA1A", "HSPA1B", "HSPA5", "HSPA6", "HSPA8", "HSPB1", "HSPD1", "HSPE1", "HSPH1")

VlnPlot(
  seurat_obj,
  features = HSP_genes,
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 6) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```


```{r fig.width=20, fig.height=15}
VlnPlot(
  seurat_obj,
  features = HSP_genes,
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 4) &
  theme(axis.text.x = element_text(angle = 60, hjust = 1, vjust = 1),
        axis.text = element_text(size=12),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```

# Stress-sensitive genes
```{r fig.width=20, fig.height=8}
hypoxiaadaptation_genes <- c("DDIT3", "XBP1", "DDIT4", "P4HB", # ER stress (Rashid et al, 2015)
                             "ENO", "HK2", "PGK1", "GAPDH", # glycolysis
                             "PDK1", "PHD", "SLC2A1" # hypoxia; https://pubmed.ncbi.nlm.nih.gov/32144406/
                             )

VlnPlot(
  seurat_obj,
  features = hypoxiaadaptation_genes,
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```

```{r fig.width=20, fig.height=15}
VlnPlot(
  seurat_obj,
  features = hypoxiaadaptation_genes,
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0, 
  ncol = 3) &
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text = element_text(size=12),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```


# SIGNATURE: Sampling-time on single-cell PBMC by Massoni-Badosa et al.
```{r fig.width=20, fig.height=8}
path_signature_Massonietal <- here::here("04_GEX_analysis/5_gene_signatures/published_signatures/signature_Massonietal_samplingtime_hPBMC.csv")
Massonietal_FULL_genes <- read_csv(path_signature_Massonietal)

Massonietal_FULLup_genes <- Massonietal_FULL_genes %>% 
  filter(direction == "up-regulated") %>% 
  dplyr::select(gene)

Massonietal_FULLup_genes_list <- Massonietal_FULLup_genes$gene
print(length(Massonietal_FULLup_genes_list))
print(Massonietal_FULLup_genes_list)
```


## Visualizing genes

```{r fig.width=14, fig.height=10}
DotPlot(seurat_obj,
        features = Massonietal_FULLup_genes_list,
        group.by = "celltypes",
        split.by = "condition",
        cols = "RdBu",
        dot.scale = 5
        ) +
   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

## Computing signature
```{r fig.width=8, fig.height=5}
# UCELL GENE SIGNATURE SCORE
## Using positive and negative gene sets, GENE+ or GENE-
seurat_obj <- AddModuleScore_UCell(seurat_obj, 
                                   features=list("Massonietal_up_genes" = Massonietal_FULLup_genes_list)
                                   )
```

## Visualizing Ucell
```{r fig.width=8, fig.height=5}
VlnPlot(
  seurat_obj,
  features = "Massonietal_up_genes_UCell",
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.format", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```

```{r fig.width=12, fig.height=8}
VlnPlot(
  seurat_obj,
  features = "Massonietal_up_genes_UCell",
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 1, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20))


ggsave(paste0(path_r_figs_out, "/HumanPBMC_MassonietalUPgenes_Ucell_Vlnbycells.svg"),
       width = 12, height = 8,
       dpi = 300)
```

```{r fig.width=10, fig.height=5}
plot_density(seurat_obj,
            features = 'Massonietal_up_genes_UCell') + 
    facet_grid(.~seurat_obj$condition)


ggsave(paste0(path_r_figs_out, "/HumanPBMC_MassonietalUPgenes_Ucelldensity.svg"),
       width = 10, height = 5,
       dpi = 300)
```


## Comparing Massonietal. signature with Fresh sDEgenes
```{r}
path_r_objects <- here::here("04_GEX_analysis/4_DE_GO_GSEA/results/R_objects")
path_object_DEgenes <- paste0(path_r_objects, "/FIXnCUT_HumanPBMC_DEgenes_Fixed_vs_Fresh.rds")
DE_genes <- readRDS(path_object_DEgenes)

DE_genes_DOWN_df <- DE_genes %>% 
    dplyr::filter(is_significant & avg_log2FC < 0) %>% 
    dplyr::select(gene, avg_log2FC, p_val_adj)
DE_genes_DOWN_list <- DE_genes_DOWN_df$gene
```

```{r fig.height=4, fig.width=4}
venndiagram <- VennDiagram::venn.diagram(
  x = list(DE_genes_DOWN_list, Massonietal_FULLup_genes_list),
  category.names = c(" ", " "),
  filename = NULL,
  fill = c("#023047", "#ffb703"),
  lty = 'blank',
  
  cex = 1.5,
  fontfamily = "sans",

  cat.cex = 1,  
  cat.fontface = "bold",
  cat.default.pos = "text",
  cat.fontfamily = "sans"
)

grid::grid.newpage()
grid::grid.draw(venndiagram)
```


```{r}
signature_genes_intersection <- intersect(Massonietal_FULLup_genes_list, DE_genes_DOWN_list)
print(signature_genes_intersection)
```


```{r fig.width=12, fig.height=9}
DotPlot(seurat_obj,
        features = signature_genes_intersection,
        group.by = "celltypes",
        split.by = "condition",
        cols = "RdBu",
        dot.scale = 5
        ) +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(axis.title = element_blank(),
        legend.position = "bottom")

ggsave(paste0(path_r_figs_out, "/HumanPBMC_MassonietalUPgenes_DEgenes.svg"),
       width = 12, height = 8,
       dpi = 300)
```



# SIGNATURE: Ex-vivo blood handling on microarrays by Baechler et al.
```{r fig.width=20, fig.height=8}
path_signature_Baechleretal <- here::here("04_GEX_analysis/5_gene_signatures/published_signatures/signature_Baechleretal_storagetime_hPBMC_microarrays.csv")
Baechleretal_FULL_genes <- read_csv(path_signature_Baechleretal)

Baechleretal_FULL_genes <- distinct(Baechleretal_FULL_genes)
Baechleretal_FULLup_genes <- Baechleretal_FULL_genes %>% 
  filter(direction == "up-regulated") %>% 
  dplyr::select(gene)

Baechleretal_FULLup_genes_list <- Baechleretal_FULLup_genes$gene
print(length(Baechleretal_FULLup_genes_list))
print(Baechleretal_FULLup_genes_list)
```


## Visualizing genes
```{r fig.width=14, fig.height=10}
DotPlot(seurat_obj,
        features = Baechleretal_FULLup_genes_list,
        group.by = "celltypes",
        split.by = "condition",
        cols = "RdBu",
        dot.scale = 5
        ) +
   theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
```

## Computing signature
```{r fig.width=8, fig.height=5}
# UCELL GENE SIGNATURE SCORE
## Using positive and negative gene sets, GENE+ or GENE-
seurat_obj <- AddModuleScore_UCell(seurat_obj, 
                                   features=list("Baechleretal_up_genes" = Baechleretal_FULLup_genes_list)
                                   )
```

## Visualizing Ucell
```{r fig.width=8, fig.height=5}
VlnPlot(
  seurat_obj,
  features = "Baechleretal_up_genes_UCell",
  group.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.format", label.y.npc = 0.95, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20)) 
```

```{r fig.width=12, fig.height=8}
VlnPlot(
  seurat_obj,
  features = "Baechleretal_up_genes_UCell",
  group.by = "celltypes",
  split.by = "condition",
  cols = palette_color_factor, 
  pt.size = 0) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median,
               geom='crossbar', colour = "black", width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test", label = "p.signif", label.y.npc = 1, label.x.npc = 0.5, size=5)&
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text = element_text(size=18),
        axis.title.x = element_blank(),
        strip.text = element_text(size=20))


ggsave(paste0(path_r_figs_out, "/HumanPBMC_BaechleretalUPgenes_Ucell_Vlnbycells.svg"),
       width = 12, height = 8,
       dpi = 300)
```

```{r fig.width=10, fig.height=5}
plot_density(seurat_obj,
            features = 'Baechleretal_up_genes_UCell') + 
    facet_grid(.~seurat_obj$condition)


ggsave(paste0(path_r_figs_out, "/HumanPBMC_BaechleretalUPgenes_Ucelldensity.svg"),
       width = 10, height = 5,
       dpi = 300)
```


## Comparing Baechleretal. signature with Fresh sDEgenes
```{r}
path_r_objects <- here::here("04_GEX_analysis/4_DE_GO_GSEA/results/R_objects")
path_object_DEgenes <- paste0(path_r_objects, "/FIXnCUT_HumanPBMC_DEgenes_Fixed_vs_Fresh.rds")
DE_genes <- readRDS(path_object_DEgenes)

DE_genes_DOWN_df <- DE_genes %>% 
    dplyr::filter(is_significant & avg_log2FC < 0) %>% 
    dplyr::select(gene, avg_log2FC, p_val_adj)
DE_genes_DOWN_list <- DE_genes_DOWN_df$gene
```

```{r fig.height=4, fig.width=4}
venndiagram <- VennDiagram::venn.diagram(
  x = list(DE_genes_DOWN_list, Baechleretal_FULLup_genes_list),
  category.names = c(" ", " "),
  filename = NULL,
  fill = c("#023047", "#ffb703"),
  lty = 'blank',
  
  cex = 1.5,
  fontfamily = "sans",

  cat.cex = 1,  
  cat.fontface = "bold",
  cat.default.pos = "text",
  cat.fontfamily = "sans"
)

grid::grid.newpage()
grid::grid.draw(venndiagram)
```


```{r}
signature_genes_intersection <- intersect(Baechleretal_FULLup_genes_list, DE_genes_DOWN_list)
print(signature_genes_intersection)
```


```{r fig.width=12, fig.height=8}
DotPlot(seurat_obj,
        features = signature_genes_intersection,
        group.by = "celltypes",
        split.by = "condition",
        cols = "RdBu",
        dot.scale = 5
        ) +
  coord_flip() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme(axis.title = element_blank(),
        legend.position = "bottom")

ggsave(paste0(path_r_figs_out, "/HumanPBMC_BaechleretalUPgenes_DEgenes.svg"),
       width = 12, height = 8,
       dpi = 300)
```


# Session Info
```{r}
sessionInfo()
```