---
title: "scRNAseq Gene Expression comparison"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
Following with the results obtained in the previous notebook, we will now perform other pre-processing steps, such as data normalization, feature selection, scaling, dimensionality reduction, and data visualization prior to batch-effect correction (data integration). To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746) and adapt some workflows from [Satija Lab](https://satijalab.org/seurat/vignettes.html).

## Objective
In this Rmarkdown document, we are going to perform the previous pre-processing steps.

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(gt)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_r_objects <- here::here("02_QC/results/R_objects")

# Functions
source(here::here("bin/utils.R"))

# Parameters
confounder_variables <- c("library_name", "sample_protocol", "fixation_time")
confounder_names <- c("Library", "Sample Protocol", "Fixation time")

metadata_variables_confounder <- c("nCount_RNA", "nFeature_RNA", "gem_id", 
                                   "pct_mt", "pct_rb", "Phase",
                                   "library_name", "sample_protocol", "fixation_time")

# Color palette
set.seed(1234) # for reproducibility
color_palette <- Polychrome::createPalette(35, c("#fc6060", "#74f774", "#7c7cfc"))
names(color_palette) <- NULL
#Polychrome::swatch(color_palette)
```

## Load data
The data used in this Rmarkdown document comes from **OMNISCOPE_39** dataset.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)

# Merging metadata tables
metadata <- metadata %>% 
  filter(subproject == "OMNISCOPE_39" & tissue == "LUNG")

print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))

# Load Seurat object
seurat_obj <- readRDS(str_c(path_r_objects, "/omniscope39_m_LUNG_healthy_stress_lognorm_harmony_processed_lowQ.rds"))
seurat_obj
DefaultAssay(seurat_obj) <- "RNA"
```


# Explore gene expression
```{r}
genes <- c("Actb", # housekeeping gene
           "Epcam", # heterogeneous epithelial cells
           "Ptprc", # Immune cells
           "Pecam1", # Endothelial cells
           "Pdgfrb" # Fibroblasts/Perivascular cells
           )
```


## General housekeeping & lineage markers
```{r fig.width=25, fig.height=6}
gg1 <- VlnPlot(
  seurat_obj,
  features = c("Actb", "Epcam", "Pecam1"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0, 
  ncol = 3) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) & # p.format)
  ylim(0, 7) &
  scale_fill_brewer(palette = "Set2")

gg2 <- VlnPlot(
  seurat_obj,
  features = c("Epcam", "Pdgfrb"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1, 
  ncol = 2) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) & 
        ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) & # p.format)
  ylim(0, 7) &
  scale_fill_brewer(palette = "Set2")

cowplot::plot_grid(plotlist = list(gg1, gg2), ncol = 2, rel_widths = c(1.5, 1))
```

```{r fig.width=10, fig.height=20}
FeaturePlot(seurat_obj,
            features = genes,
            split.by = "sample_protocol",
            pt.size = 0.25,
            ncol = 5) &
  scale_colour_gradientn(colours = viridis::inferno(10))
```


## Low quality genes
```{r}
lowquality_genes <- c("Hbb-bs", "Hba-a1", "Hba-a2", "Hbb-bt")
```

```{r fig.width=20, fig.height=6}
VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "sample_protocol",
        #split.by = "sample_protocol", 
        ncol = 4)  &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95, size=5) & # p.format)
  ylim(0, 9) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=20, fig.height=6}
VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "sample_protocol",
        ncol = 4)  &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95, size=5) & # p.format)
  ylim(0, 9) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=10, fig.height=12}
FeaturePlot(seurat_obj,
            features = c("Hbb-bs", "Hba-a1", "Hba-a2", "Hbb-bt"),
            split.by = "sample_protocol",
            pt.size = 0.25,
            order = T,
            ncol = 5) &
  scale_colour_gradientn(colours = viridis::inferno(10))
```


# Massoni et al. gene signatures
```{r}
TOP_UPgenes <- c("Nfkbia", "Eif1", "Jund", "Jun", "Ubc", "Jundb", "Tnfaip3", "Cdc42", "Id3", "Fosb", "Ppp1r15a",
                          "Lefprot1", "Calr", "Hspa1a", "Hspa1b", "Hspa8", "Trac", "Stx7", "Sit1", "Sec62")
```

```{r fig.width=20, fig.height=20}
VlnPlot(
  seurat_obj,
  features = c("Actb", TOP_UPgenes),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1, 
  ncol = 5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
        ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=17, fig.height=5}
DotPlot(seurat_obj,
        features = c("Actb", TOP_UPgenes),
        group.by = "sample_protocol",
        dot.scale = 10,
        )
```


```{r fig.width=8, fig.height=5}
# Gene signature associated to collagenase treatment in mouse satellite cells

vandenBrink_signature <- c("Cebpb", "Id3", "Jund", "Hspb1", "Dusp1", "Dnajb1", "Fos", "Hspa8",
                           "Atf3", "Mt1m", "Hsp90ab1", "Junb", "Hsph1", "Jun", "Hspa1a", "Hspa1b", "Nfkbia",
                           "Egr1", "Fosb", "Cebpd", "Socs3", "Nr4a1", "Ubc", "Zfp36", "Ppp1r15a", "Btg2", "Ier2")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(vandenBrink_signature),
    name = "vandenBrink_signature"
)

VlnPlot(
  seurat_obj,
  features = "vandenBrink_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1,
  ncol = 1) &
  ylim(0, 1.5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=8, fig.height=5}
# Gene signature associated to storage time in PBMC (measured by microarrays)

Baechler_signature <- c("Id1", "Cdc42", "Phlda2", "Nr4a2", "Jun", "Mir22hg", "Nr4a2", "Mica", "Adm", 
                        "Fosb", "Cd83", "Klf10", "Dusp2", "Ppp1r15a", "Ctrl", "Nfil3", "Nfkbia", "Plk2", "Ptx3",
                        "Gadd45b", "Vegfa", "Arl4a", "Cdkn1a", "Junb", "Tnfaip3", "Eif1")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(Baechler_signature),
    name = "Baechler_signature"
)

VlnPlot(
  seurat_obj,
  features = "Baechler_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1,
  ncol = 1) &
  ylim(0, 1) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=8, fig.height=5}
# Gene signature associated to storage time in PBMC (measured by scRNA-seq)

PBMC_Massoni_signature <- c("Fth1", "Eif1", "Crem", "H3f3b", "Leprotl1", "Gabarapl1", "Arid5a",
                            "Hmgb2", "Selk", "Cirbp", "Cnot1", "Cxcr4", "Znf331", "Rnf145", "Nfkbia",
                            "Cib1", "Syap1", "Chd1", "Pik3r1", "Fam177a1", "G0s2", "Neat1", "Fam133b",
                            "Sbds", "Jund", "Stat4", "Kmt2e", "Eif1b", "Tspyl2", "Tnfaip3", "Mgat4a",
                            "Sytl3", "Ifngr1", "Spock2", "Ezr", "Cd44")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(PBMC_Massoni_signature),
    name = "PBMC_Massoni_signature"
)

VlnPlot(
  seurat_obj,
  features = "PBMC_Massoni_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1,
  ncol = 1) &
  ylim(0, 0.75) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95, size=5) &
  scale_fill_brewer(palette = "Set2")
```


# Session Info
```{r}
sessionInfo()
```