---
title: "scRNAseq Gene Expression comparison"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction
Following with the results obtained in the previous notebook, we will now perform other pre-processing steps, such as data normalization, feature selection, scaling, dimensionality reduction, and data visualization prior to batch-effect correction (data integration). To do so, we will follow the current best practices for scRNA-seq QC described in [Luecken al. Mol Syst Biol (2018)](https://doi.org/10.15252/msb.20188746) and adapt some workflows from [Satija Lab](https://satijalab.org/seurat/vignettes.html).

## Objective
In this Rmarkdown document, we are going to perform the previous pre-processing steps.

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(gt)
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")
path_r_objects <- here::here("03_clustering_annotation/results/R_objects")

# Functions
source(here::here("bin/utils.R"))

# Parameters
confounder_variables <- c("library_name", "sample_protocol", "fixation_time")
confounder_names <- c("Library", "Sample Protocol", "Fixation time")

metadata_variables_confounder <- c("nCount_RNA", "nFeature_RNA", "gem_id", 
                                   "pct_mt", "pct_rb", "Phase",
                                   "library_name", "sample_protocol", "fixation_time")

# Color palette
set.seed(1234) # for reproducibility
color_palette <- Polychrome::createPalette(35, c("#fc6060", "#74f774", "#7c7cfc"))
names(color_palette) <- NULL
#Polychrome::swatch(color_palette)
```

## Load data
The data used in this Rmarkdown document comes from **SCGTEST_34** dataset.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)

# Merging metadata tables
metadata <- metadata %>% 
  filter(subproject == "SCGTEST_34" & tissue == "PBMC")

print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))

# Load Seurat object
seurat_obj <- readRDS(str_c(path_r_objects, "/scgtest34_h_PBMC_healthy_cells1_clustering_annotation.rds"))
seurat_obj
```

```{r}
# Remove doublets
seurat_obj <- seurat_obj[, seurat_obj$celltypes_1 != "Low quality cells"]
```

## Cell proportions
```{r fig.width=12, fig.height=6}
gg_cell_num <- dittoSeq::dittoBarPlot(seurat_obj, 
                            "celltypes_1",
                            group.by = "sample_protocol",
                            scale = "count",
                            retain.factor.levels = TRUE,
                            legend.show = FALSE,
                            color.panel = color_palette)

gg_cell_prop <- dittoSeq::dittoBarPlot(seurat_obj, 
                             "celltypes_1",
                             group.by = "sample_protocol",
                             scale = "percent",
                            retain.factor.levels = TRUE,
                             color.panel = color_palette)

gg_cell_num + gg_cell_prop
```

# Explore gene expression
```{r}
genes <- c("ACTB", # housekeeping gene
           "EPCAM", # heterogeneous epithelial cells
           "PTPRC", # Immune cells
           "PECAM1", # Endothelial cells
           "PDGFRB" # Fibroblasts/Perivascular cells
           )
```


## General housekeeping & lineage markers
```{r fig.width=25, fig.height=6}
gg1 <- VlnPlot(
  seurat_obj,
  features = c("ACTB", "EPCAM", "PECAM1"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0, 
  ncol = 3) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) & # p.format)
  ylim(0, 7) &
  scale_fill_brewer(palette = "Set2")

gg2 <- VlnPlot(
  seurat_obj,
  features = c("EPCAM", "PDGFRB"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0.1, 
  ncol = 2) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) & 
        ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) & # p.format)
  ylim(0, 7) &
  scale_fill_brewer(palette = "Set2")

cowplot::plot_grid(plotlist = list(gg1, gg2), ncol = 2, rel_widths = c(1.5, 1))
```

```{r fig.width=10, fig.height=20}
FeaturePlot(seurat_obj,
            features = genes,
            split.by = "sample_protocol",
            pt.size = 0.25,
            ncol = 5) &
  scale_colour_gradientn(colours = viridis::inferno(10))
```


## Low quality genes
```{r}
lowquality_genes <- c("HBB", "HBA1", "HBA2")
```

```{r fig.width=20, fig.height=6}
VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "sample_protocol",
        #split.by = "sample_protocol", 
        ncol = 4)  &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95) & # p.format)
  ylim(0, 9) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=20, fig.height=6}
VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "celltypes_1",
        split.by = "sample_protocol", 
        ncol = 4)  &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95) & # p.format)
  ylim(0, 9) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=10, fig.height=12}
FeaturePlot(seurat_obj,
            features = c("HBB", "HBA1", "HBA2"),
            split.by = "sample_protocol",
            pt.size = 0.25,
            order = T,
            ncol = 5) &
  scale_colour_gradientn(colours = viridis::inferno(10))
```


```{r fig.width=20, fig.height=5}
VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "sample_protocol",
        idents = "Pheumocytes Type I",
        pt.size = 0.1, 
        ncol = 4) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95) & # p.format)
  ylim(0, 8.5) &
  scale_fill_brewer(palette = "Set2")

VlnPlot(seurat_obj,
        features = lowquality_genes,
        group.by = "sample_protocol",
        idents = "Pneumocyt es Type II",
        pt.size = 0.1, 
        ncol = 4) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
      ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.signif", label.y.npc = 0.95) & # p.format)
  ylim(0, 8.5) &
  scale_fill_brewer(palette = "Set2")
```

# Massoni et al. gene signatures
```{r}
TOP_UPgenes <- c("NFKBIA", "EIF1", "JUND", "JUN", "UBC", "JUNDB", "TNFAIP3", "CDC42", "ID3", "FOSB", "PPP1R15A",
                          "LEFPROT1", "CALR", "HSPA1A", "HSPA1B", "HSPA8", "TRAC", "STX7", "SIT1", "SEC62")
```

```{r fig.width=20, fig.height=20}
VlnPlot(
  seurat_obj,
  features = c("ACTB", TOP_UPgenes),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0, 
  ncol = 5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) & # p.format)  
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=17, fig.height=5}
DotPlot(seurat_obj,
        features = c("ACTB", TOP_UPgenes),
        group.by = "sample_protocol",
        dot.scale = 10,
        )
```


```{r fig.width=8, fig.height=5}
# Gene signature associated to collagenase treatment in mouse satellite cells

vandenBrink_signature <- c("CEBPB", "ID3", "JUND", "HSPB1", "DUSP1", "DNAJB1", "FOS", "HSPA8",
                           "ATF3", "MT1M", "HSP90AB1", "JUNB", "HSPH1", "JUN", "HSPA1A", "HSPA1B", "NFKBIA",
                           "EGR1", "FOSB", "CEBPD", "SOCS3", "NR4A1", "UBC", "ZFP36", "PPP1R15A", "BTG2", "IER2")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(vandenBrink_signature),
    name = "vandenBrink_signature"
)

VlnPlot(
  seurat_obj,
  features = "vandenBrink_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 1.5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=8, fig.height=5}
# Gene signature associated to storage time in PBMC (measured by microarrays)

Baechler_signature <- c("ID1", "CDC42", "PHLDA2", "NR4A2", "JUN", "MIR22HG", "NR4A2", "MICA", "ADM", 
                        "FOSB", "CD83", "KLF10", "DUSP2", "PPP1R15A", "CTRL", "NFIL3", "NFKBIA", "PLK2", "PTX3",
                        "GADD45B", "VEGFA", "ARL4A", "CDKN1A", "JUNB", "TNFAIP3", "EIF1")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(Baechler_signature),
    name = "Baechler_signature"
)

VlnPlot(
  seurat_obj,
  features = "Baechler_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 1) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")
```

```{r fig.width=8, fig.height=5}
# Gene signature associated to storage time in PBMC (measured by scRNA-seq)

PBMC_Massoni_signature <- c("FTH1", "EIF1", "CREM", "H3F3B", "LEPROTL1", "GABARAPL1", "ARID5A",
                            "HMGB2", "SELK", "CIRBP", "CNOT1", "CXCR4", "ZNF331", "RNF145", "NFKBIA",
                            "CIB1", "SYAP1", "CHD1", "PIK3R1", "FAM177A1", "G0S2", "NEAT1", "FAM133B",
                            "SBDS", "JUND", "STAT4", "KMT2E", "EIF1B", "TSPYL2", "TNFAIP3", "MGAT4A",
                            "SYTL3", "IFNGR1", "SPOCK2", "EZR", "CD44")

seurat_obj <- AddModuleScore(
    seurat_obj,
    features = list(PBMC_Massoni_signature),
    name = "PBMC_Massoni_signature"
)

VlnPlot(
  seurat_obj,
  features = "PBMC_Massoni_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 0.75) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")
```

# Cell-type signatures
```{r}
housekeeping_signature <- c("ACTB", "B2M",  "GUSB", "HMBS", "RPL13A", "SDHA")
Bcell_signature <- c("CD79A", "CD79B", "MS4A1", "CD19", "CD74", "LY6D", "IGHM", "IGKC")
Tcell_signature <- c("CD3D", "CD3E", "CD4", "CD8A", "NKG7", "IRF7")
Myeloid_signature <- c("S100A4", "CD14", "LYZ", "CCR2", "PLAC8")

signature_list <- list(housekeeping_signature, Bcell_signature, Tcell_signature, Myeloid_signature)
names(signature_list) <- c("housekeeping_signature", "Bcell_signature", "Tcell_signature", "Myeloid_signature")
  
seurat_obj <- AddModuleScore(
    seurat_obj,
    features = signature_list,
    name = c("housekeeping_signature", "Bcell_signature", "Tcell_signature", "Myeloid_signature")
)
```

## VlnPlot
```{r fig.width=8, fig.height=5}
VlnPlot(
  seurat_obj,
  features = "housekeeping_signature1",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 2.5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")

VlnPlot(
  seurat_obj,
  features = "Bcell_signature2",
  idents = "B cells",
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 4) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")


VlnPlot(
  seurat_obj,
  features = "Tcell_signature3",
  idents = c("CD8+ T cells", "CM CD4+ T cells 1", "CM CD4+ T cells 2", "Cytotoxic CD8+ T cells", 
             "Naive CD4+ T cells", "Naive CD8+ T cells"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 2) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")

VlnPlot(
  seurat_obj,
  features = "Myeloid_signature4",
  idents = c("CD14+ Monocytes", "DC2", "DC4"),
  group.by = "sample_protocol",
  cols = color_palette, 
  pt.size = 0,
  ncol = 1) &
  ylim(0, 2.5) &
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom='crossbar', colour = "black",
                 width=0.25, position = position_dodge(width = .25)) &
  ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
                                 label = "p.format", label.y.npc = 0.95) &
  scale_fill_brewer(palette = "Set2")
```


## RidgePlot
```{r fig.width=10, fig.height=6}
RidgePlot(seurat_obj,
          features = "housekeeping_signature1",
  group.by = "sample_protocol")

RidgePlot(seurat_obj,
          features = "Bcell_signature2",
  idents = c("B cells"),
  group.by = "sample_protocol")

RidgePlot(seurat_obj,
          features = "Tcell_signature3",
  idents = c("CD8+ T cells", "CM CD4+ T cells 1", "CM CD4+ T cells 2", "Cytotoxic CD8+ T cells", 
             "Naive CD4+ T cells", "Naive CD8+ T cells"),
  group.by = "sample_protocol")

RidgePlot(seurat_obj,
          features = "Myeloid_signature4",
  idents = c("CD14+ Monocytes", "DC2", "DC4"),
  group.by = "sample_protocol")
```



# Session Info
```{r}
sessionInfo()
```