---
title: "scRNAseq QC parameters"
author: "Laura Jim√©nez Gracia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE,
                      fig.align = "center")
```

# Introduction

## Objective
In this Rmarkdown document, we are going to 

# Pre-processing
## Libraries
```{r warning = FALSE, message = FALSE}
library(tidyverse)
library(Seurat)
library(gt)
library(scales)
library(DropletUtils)
library(scuttle)
library(Rcpp)
sourceCpp("downsample_run.cpp")
```

## Parameters
Here we will define and load objects and functions that will be used throughout the document.
```{r}
# Paths
path_project_metadata <- here::here("01_cellranger_mapping/data/FIXnCUT_metadata.csv")

# Functions
source(here::here("bin/utils.R"))

set.seed(1234) # for reproducibility 

palette_color_factor <- c("Fresh" = "#66C2A5", 
                          "Fixed" ="#FC8D62")
```

## Load metadata
The data used in this Rmarkdown document comes from **SCGTEST_33** dataset.
```{r}
# Load all metadata information
metadata <- read.csv(path_project_metadata)

# Merging metadata tables
metadata <- metadata %>% 
  filter(subproject == "SCGTEST_33"
         & tissue == "Lung"
         & ( sample_protocol == "Fresh" | sample_protocol == "Fix&Cut_24h"))


# Modify metadata
metadata$library_name[metadata$library_name == "Lung_fresh"] <- "Fresh"
metadata$library_name[metadata$library_name == "Lung_fixed_24h"] <- "Fixed"

metadata$sample_protocol <- metadata$library_name


print("Libraries/Samples metadata")
DT::datatable(metadata, options = list(scrollX = TRUE))
```


# CELLRANGER OUTPUT -- Raw: Assessing sequencing depth

https://github.com/MarioniLab/DropletUtils/blob/master/vignettes/DropletUtils.Rmd

Downsampling on the reads: Given multiple batches of very different sequencing depths, it can be beneficial to downsample the deepest batches to match the coverage of the shallowest batches. This avoids differences in technical noise that can drive clustering by batch.

The `r Biocpkg("scuttle")` package provides some utilities to downsample count matrices, but technically speaking, downsampling on the reads is more appropriate as it recapitulates the effect of differences in sequencing depth per cell.

This can be achieved by applying the `downsampleReads` function to the molecule information file containing the read counts. Users should use `downsampleMatrix()` instead if they want to guarantee similar total counts after downsampling.


```{r}
set.seed(0123)

seq_list <- c(10000000, 20000000, 30000000, 40000000, 50000000, 60000000, 70000000, 80000000,
              90000000, 100000000, 110000000, 120000000)
seqreads_genes_df <- data.frame(matrix(ncol = 4, nrow = 0))
colnames(seqreads_genes_df) <- c("sequenced_reads", "detected_genes", "detected_umis", "sample_protocol")

for (protocol in metadata$sample_protocol) {
  print(protocol)
  gem_id <- unique(metadata$gem_id[metadata$sample_protocol == protocol])
  path_general <- here::here("01_cellranger_mapping/subprojects/SCGTEST_33/jobs")
  mol.info.path <- paste(path_general, gem_id, gem_id, "outs/per_sample_outs", gem_id, "count/sample_molecule_info.h5", sep = "/")
  mol.info.file <- read10xMolInfo(mol.info.path)
  
  all_reads <- sum(mol.info.file$data$reads)
  print(all_reads)
  prop_list <- seq_list / all_reads

  for (prop in prop_list) {
    
    sequenced_reads <- all_reads*prop
    downsampled_reads <- downsample_run(mol.info.file$data$reads, prop = prop)
    detected_genes <- length(unique(mol.info.file$data$gene[downsampled_reads>0]))
    detected_umis <- sum(downsampled_reads>0) # == nrow(unique(mol.info.file$data[downsampled_reads>0, c("cell", "umi")]))
    
    seqreads_genes_df[nrow(seqreads_genes_df)+1, ] <- c(sequenced_reads, detected_genes, detected_umis, protocol)
  }
}

seqreads_genes_df$sequenced_reads <- as.integer(seqreads_genes_df$sequenced_reads)
seqreads_genes_df$detected_genes <- as.integer(seqreads_genes_df$detected_genes)
seqreads_genes_df$detected_umis <- as.integer(seqreads_genes_df$detected_umis)

seqreads_genes_df$sample_protocol[seqreads_genes_df$sample_protocol == "Fix&Cut_24h"] <- "Fixed"

head(seqreads_genes_df)
write_csv(x = seqreads_genes_df,
          file = "tables/MouseLung_fixed_perfused_LibGenesUMIvsReads.csv")
```


## # Genes / Seq reads
```{r fig.width=8, fig.height=5}
seqreads_genes_df <- read.csv("tables/MouseLung_fixed_perfused_LibGenesUMIvsReads.csv")

gg_genes <- seqreads_genes_df %>%
  ggplot(aes(x = sequenced_reads, y = detected_genes, color = sample_protocol)) + 
  geom_point(size=4) +
  geom_smooth(se = FALSE) +
  #scale_y_continuous(labels = scales::comma) + 
  scale_x_continuous(labels = label_number(scale = 1e-6), breaks = seq(0, 130000000, by = 20000000)) +
  theme_classic() +
  scale_color_manual(values = palette_color_factor) +
  scale_fill_manual(values = palette_color_factor) +
  labs(x = "",
     y = "# Detected Genes",
     color = "Protocol") +
    theme(axis.title = element_text(size = 18),
          axis.text = element_text(size = 14),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),          
          legend.position = c(0.9, 0.2),
          legend.text = element_text(size = 16),
          legend.title=element_blank())


gg_umis <- seqreads_genes_df %>%
  ggplot(aes(x = sequenced_reads, y = detected_umis, color = sample_protocol)) + 
  geom_point(size=4) +
  geom_smooth(se = FALSE) +
  scale_y_continuous(labels = label_number(scale = 1)) + 
  scale_x_continuous(labels = label_number(scale = 1e-6), breaks = seq(0, 130000000, by = 20000000)) +
  theme_classic() +
  scale_color_manual(values = palette_color_factor) +
  scale_fill_manual(values = palette_color_factor) +
  labs(x = "# Sequencing Reads (in millions)",
     y = "# Detected UMIs",
     color = "Protocol") +
    theme(axis.title = element_text(size = 18),
          axis.text = element_text(size = 14),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = "none")


gg <- gridExtra::grid.arrange(gg_genes, gg_umis, ncol = 1)

ggsave(plot=gg,  filename = "figs/MouseLung_fixed_perfused_LibGenesUMIvsReads.svg",
       width = 8, height = 5,
       dpi = 300)
```

## Statistics
```{r}
seqreads_genes_df <- read.csv("tables/MouseLung_fixed_perfused_LibGenesUMIvsReads.csv")

## (1) statistical tests
# Perform a Mann-Whitney U test for genes between Fresh and Fixed samples
wilcoxon_genes <- wilcox.test(detected_genes ~ sample_protocol, data = seqreads_genes_df)
print(wilcoxon_genes)
#print(wilcoxon_genes$p.value)

# Perform a Mann-Whitney U test for UMIs between Fresh and Fixed samples
wilcoxon_umis <- wilcox.test(detected_umis ~ sample_protocol, data = seqreads_genes_df)
print(wilcoxon_umis)
#print(wilcoxon_umis$p.value)



## (2) fit a regression model & if the confidence intervals overlap, suggesting no significant difference.
# Fit linear regression models
model_fresh_genes <- lm(detected_genes ~ sequenced_reads, data = subset(seqreads_genes_df, sample_protocol == "Fresh"))
model_fixed_genes <- lm(detected_genes ~ sequenced_reads, data = subset(seqreads_genes_df, sample_protocol == "Fixed"))

# Obtain confidence intervals for the coefficient of X
ci_model_fresh_genes<- confint(model_fresh_genes)["sequenced_reads", ]
ci_model_fixed_genes <- confint(model_fixed_genes)["sequenced_reads", ]
print("detected_genes")
print(ci_model_fresh_genes)
print(ci_model_fixed_genes)

# Check if confidence intervals overlap
ci_overlap_genes <- ci_model_fresh_genes[2] > ci_model_fixed_genes[1] && ci_model_fixed_genes[2] > ci_model_fresh_genes[1]
ci_overlap_genes


# Fit linear regression models
model_fresh_umis <- lm(detected_umis ~ sequenced_reads, data = subset(seqreads_genes_df, sample_protocol == "Fresh"))
model_fixed_umis <- lm(detected_umis ~ sequenced_reads, data = subset(seqreads_genes_df, sample_protocol == "Fixed"))

# Obtain confidence intervals for the coefficient of X
ci_model_fresh_umis <- confint(model_fresh_umis)["sequenced_reads", ]
ci_model_fixed_umis <- confint(model_fixed_umis)["sequenced_reads", ]
print("detected_umis")
print(ci_model_fresh_umis)
print(ci_model_fixed_umis)

# Check if confidence intervals overlap
ci_overlap_umis <- ci_model_fresh_umis[2] > ci_model_fixed_umis[1] && ci_model_fixed_umis[2] > ci_model_fresh_umis[1]
ci_overlap_umis
```


# CELLRANGER OUTPUT -- Unfiltered
```{r}
cell_info_df <- data.frame(matrix(ncol = 7, nrow = 0))
colnames(cell_info_df) <- c("cell", "reads", "umi", "gene", "library")


for (library in metadata$library_name) {
  #print(protocol)
  gem_id <- unique(metadata$gem_id[metadata$library_name == library])
  subproject <- unique(metadata$subproject[metadata$library_name == library])
  protocol <- unique(metadata$sample_protocol[metadata$library_name == library])
  replicate <- unique(metadata$replicate[metadata$library_name == library])

  path_general <- here::here(paste0("01_cellranger_mapping/subprojects/", subproject, "/jobs"))
  mol.info.path <- paste(path_general, gem_id, gem_id, "outs/per_sample_outs", gem_id, "count/sample_molecule_info.h5", sep = "/")
  mol.info.file <- DropletUtils::read10xMolInfo(mol.info.path)
  
    # Get the number of reads / cell [Category, x]
  reads_cell_df <- aggregate(mol.info.file$data$reads, 
                             by= list(cell=(mol.info.file$data$cell)),
                             FUN=sum)
  names(reads_cell_df)[names(reads_cell_df) == "x"] <- "reads"
  
  # Get the number of umi / cell [Category, x]
  umi_cell_df <- aggregate(data = mol.info.file$data,
            umi ~ cell, 
            FUN=function(umi) length(unique(umi))
            )
  
  # Get the number of genes / cell [Category, x]
  gene_cell_df <- aggregate(data = mol.info.file$data,
            gene ~ cell, 
            FUN=function(gene) length(unique(gene))
            )
  
  df <- full_join(reads_cell_df, umi_cell_df, by = "cell")
  df <- full_join(df, gene_cell_df, by = "cell")
  df$library <- library
  df$sample_protocol <- protocol
  df$replicate <- replicate
  

  # Information
  #sum(df$reads)
  #sum(df$umi)
  #No possible for genes, as they might be repeated across cells!
  
  cell_info_df <- rbind(cell_info_df, df)
}

head(cell_info_df)
```


```{r}
# Computing slope genes
for (library in metadata$library_name) {
  fit1 <- lm(formula =  log10(gene) ~ log10(reads), data = cell_info_df[cell_info_df$library == library, ])
  #summary(fit1)
  print(library)
  slope <- round(coef(fit1)[2], digits = 4)
  print(slope)

  index <- rownames(cell_info_df[cell_info_df$library == library, ][which.max(cell_info_df[cell_info_df$library == library, ]$reads), ])
  cell_info_df[index, "slope_gene"] <- slope
}

# Computing slope UMIs
for (library in metadata$library_name) {
  fit1 <- lm(formula =  log10(umi) ~ log10(reads), data = cell_info_df[cell_info_df$library == library, ])
  #summary(fit1)
  print(library)
  slope <- round(coef(fit1)[2], digits = 4)
  print(slope)

  index <- rownames(cell_info_df[cell_info_df$library == library, ][which.max(cell_info_df[cell_info_df$library == library, ]$reads), ])
  cell_info_df[index, "slope_umi"] <- slope
}

head(cell_info_df)
write_csv(x = cell_info_df,
          file = "tables/MouseLung_fixed_perfused_LibGenes-UMIsvsReadsCell.csv")
```


## # Genes / # Seq reads (per cell)
```{r fig.width=8, fig.height=5}
cell_info_df <- read.csv("tables/MouseLung_fixed_perfused_LibGenes-UMIsvsReadsCell.csv")

gg <- cell_info_df %>%
  ggplot(aes(x = reads, y = gene, color = library, fill = library)) +
  geom_point(alpha=0.1) + 
  #geom_rug() +
  scale_color_manual(values = palette_color_factor) +
  scale_fill_manual(values = palette_color_factor) +
  stat_smooth(method='lm', se=TRUE) + # fitted linear regression model
    scale_y_log10(labels = scales::comma) +
  scale_x_log10(labels = scales::comma) +
  labs(x = "# Sequencing Reads",
     y = "# Detected Genes",
     color = "Protocol") +
  theme_classic() +
    theme(axis.title = element_text(size = 16),
          axis.text = element_text(size = 12),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = c(0.85, 0.25),
          legend.text = element_text(size = 14),
          legend.title=element_blank()) +
    guides(fill = guide_legend(override.aes = list(color = NA), reverse = TRUE), 
           color = FALSE
           )
  
ggExtra::ggMarginal(gg, groupColour = TRUE, groupFill = TRUE)

ggsave("figs/MouseLung_fixed_perfused_LibGenesvsReadsCell.svg",
       width = 8, height = 5,
       dpi = 300)
```

## # UMIs / # Seq reads (per cell)

```{r fig.width=8, fig.height=5}
cell_info_df <- read.csv("tables/MouseLung_fixed_perfused_LibGenes-UMIsvsReadsCell.csv")

gg <- cell_info_df %>%
  ggplot(aes(x = reads, y = umi, color = library, fill = library)) +
  geom_point(alpha=0.1) + 
  #geom_rug() +
  scale_color_manual(values = palette_color_factor) +
  scale_fill_manual(values = palette_color_factor) +
  stat_smooth(method='lm', se=TRUE) + # fitted linear regression model
  scale_x_log10(labels = scales::comma) +
  scale_y_log10(labels = scales::comma) +
  labs(x = "# Sequencing Reads",
     y = "# Detected UMIs",
     color = "Protocol") +
  theme_classic() +
    theme(axis.title = element_text(size = 16),
          axis.text = element_text(size = 12),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = c(0.85, 0.25),
          legend.text = element_text(size = 14),
          legend.title=element_blank()) +
    guides(fill = guide_legend(override.aes = list(color = NA), reverse = TRUE), 
           color = FALSE
           )
ggExtra::ggMarginal(gg, groupColour = TRUE, groupFill = TRUE)

ggsave("figs/MouseLung_fixed_perfused_LibUMIsvsReadsCell.svg",
       width = 8, height = 5,
       dpi = 300)
```


# CELLRANGER OUTPUT -- Filtered (Cellranger)
```{r}
# Load Seurat object
seurat_obj <- readRDS(here::here("02_QC/results/R_objects/FIXnCUT_MouseLung_fixed_perfused_merged.rds"))
seurat_obj
```


```{r}
size_list <- seq(1, 100, by=2)
seurat_obj$library_name <- as.factor(seurat_obj$library_name)

detectedgenes_df <- data.frame(matrix(
  ncol =  length(levels(seurat_obj$library_name)),
  nrow = length(size_list)), 
  row.names = as.character(size_list))
names(detectedgenes_df) <- levels(seurat_obj$library_name)


Idents(seurat_obj) <- "library_name"
for (size in size_list) {
  #print(size)
  for (library in levels(seurat_obj$library_name)) {
  #print(library)
    seurat_obj_sub <- seurat_obj[, seurat_obj$library_name == library]
    detectedgenes_list <- list()

    for (var in 1:50) {
      seurat_obj_sub_n <- subset(x = seurat_obj_sub, downsample = size)
      #print(table(seurat_obj_sub$sample_protocol))
  
      m <- rowSums(seurat_obj_sub_n@assays$RNA@counts != 0)
      detected_genes <- length(m[m > 0])
      #print(detected_genes)
      detectedgenes_list[var] <- detected_genes
    }
  
    mean_detectedgenes <- mean(unlist(detectedgenes_list))
    #print(mean_detectedgenes)
    detectedgenes_df[as.character(size), library] <- mean_detectedgenes  
    
  }
}

detectedgenes_df
detectedgenes_df$num_cells <- as.integer(rownames(detectedgenes_df))

detectedgenes <- detectedgenes_df %>% 
  gather(key = "library_name", value = "detected_genes", `Fixed`:`Fresh`)

detectedgenes$sample_protocol <- detectedgenes$library_name

head(detectedgenes)
write_csv(x = detectedgenes,
          file = "tables/MouseLung_fixed_perfused_CumulativeGenes.csv")
```


## Cumulative # Genes / # cells
```{r fig.width=8, fig.height=5}
detectedgenes <- read.csv("tables/MouseLung_fixed_perfused_CumulativeGenes.csv")

detectedgenes %>% 
  ggplot(aes(x = num_cells, y = detected_genes, color = sample_protocol)) +
  geom_point(size=3) +
  scale_color_manual(values = palette_color_factor) +
  scale_fill_manual(values = palette_color_factor) +
  labs(x = "# Sampled Cells",
     y = "# Detected Genes") +
  theme_classic() +
  theme(axis.title = element_text(size = 16),
          axis.text = element_text(size = 12),
          strip.placement = "outside",
          strip.background = element_rect(colour = NA),
          legend.position = c(0.9, 0.2),
          legend.text = element_text(size = 14),
          legend.title=element_blank()) +
    guides(color = guide_legend(override.aes = list(fill = NA), reverse = TRUE))

ggsave("figs/MouseLung_fixed_perfused_CumulativeGenes.svg",
       width = 8, height = 5,
       dpi = 300)
```


## QC metrics
Here, we show an overview of the scRNA-seq data obtained after the quality control.
```{r}
table_qc_gex(seurat_obj@meta.data, subtitle = "After cell QC filtering out")
```


```{r fig.height=6, fig.width=20}
VlnPlot(seurat_obj,
        features = c("nCount_RNA", "nFeature_RNA", "pct_mt", "pct_rb"),
        group.by = "sample_protocol",
        pt.size = 0.01,
        log = TRUE,
        ncol = 4
        ) &
    scale_fill_brewer(palette = "Set2") &
    theme(axis.text.x = element_text(angle = 0, hjust = 0.5),
          axis.title = element_blank())
  #theme_classic() &
  #ggpubr::stat_compare_means(ref.group = "Fresh", method = "wilcox.test",
  #                               label = "p.format", label.y.npc = 0.95)
```


# Session Info
```{r}
sessionInfo()
```